{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "INFO: Pandarallel will run on 10 workers.\n",
      "INFO: Pandarallel will use standard multiprocessing data transfer (pipe) to transfer data between the main process and workers.\n"
     ]
    }
   ],
   "source": [
    "## Import statements\n",
    "import pandas as pd\n",
    "import baymobil as baymob\n",
    "from pandarallel import pandarallel\n",
    "pandarallel.initialize(progress_bar = False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [],
   "source": [
    "## Define functions\n",
    "## Function to calculating Bayes Factors for the Thieme data, based on a set error rate (the error rate is defined as 1 in \"error\")\n",
    "\n",
    "def calculate_bf(row, error):\n",
    "        \"\"\"\n",
    "        Calculate the Bayes Factor (BF) for a given row.\n",
    "        The function handles cases where N and n values may be in comma-separated format.\n",
    "        \"\"\"\n",
    "        bf_list = []\n",
    "\n",
    "        # Separate into integers and handle multiple values\n",
    "        endo_list, distal_list = [], []\n",
    "        \n",
    "        if \",\" in str(row.endo):\n",
    "            endo_list = [int(n) for n in row.endo.split(\",\") if \"NA\" not in n]\n",
    "            distal_list = [int(n) for n in row.distal.split(\",\") if \"NA\" not in n]\n",
    "        else:\n",
    "            try:\n",
    "                endo_list = [int(row.endo)]\n",
    "                distal_list = [int(row.distal)]\n",
    "            except ValueError:\n",
    "               # print(\"NAN encountered in row:\", row)\n",
    "                return \"NA\"\n",
    "        \n",
    "        # Calculate the Bayes Factor for each pair of N and n\n",
    "        for endo, distal in zip(endo_list, distal_list):\n",
    "            N = endo + distal\n",
    "            n = distal\n",
    "            bf = baymob.fasterpostN2(error, 1, error, 1, N, n, 10)\n",
    "            bf_value = min(max(bf[2], -2), 2)  # Cap the BF values between -2 and 2\n",
    "            bf_list.append(bf_value)\n",
    "\n",
    "        return sum(bf_list)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/Users/tomkinsm/miniconda3/lib/python3.12/site-packages/openpyxl/worksheet/_reader.py:329: UserWarning: Unknown extension is not supported and will be removed\n",
      "  warn(msg)\n"
     ]
    }
   ],
   "source": [
    "## Load in the Excel table from the supplementary material: https://www.nature.com/articles/nplants201525#Sec15 - the cell runs in ~ 14 secs.\n",
    "df = pd.read_excel(\"../Data/thieme_original_data.xlsx\", sheet_name=\"SNP in all transcripts\", header=[0, 1])\n",
    "\n",
    "# Merge the two levels of the header into a single level with a more meaningful separator\n",
    "df.columns = [':'.join(col).strip() for col in df.columns.values]\n",
    "\n",
    "## Drop the extra columns from the dataframe\n",
    "df = df.drop(columns=df.filter(like=\"Status.\").columns).copy()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>PrCs* FN_stemUpper:PED</th>\n",
       "      <th>PrCs* FN_stemUpper:COL</th>\n",
       "      <th>PrCs* FN_stemUpper:Status</th>\n",
       "      <th>CrPs* -P r1:PED</th>\n",
       "      <th>CrPs* -P r1:COL</th>\n",
       "      <th>CrPs* -P r1:Status</th>\n",
       "      <th>Pr*Cs FN_root:PED</th>\n",
       "      <th>Pr*Cs FN_root:COL</th>\n",
       "      <th>Pr*Cs FN_root:Status</th>\n",
       "      <th>Cr*Ps -N r2:PED</th>\n",
       "      <th>...</th>\n",
       "      <th>Pr*Cs r2:PED</th>\n",
       "      <th>Pr*Cs r2:COL</th>\n",
       "      <th>Pr*Cs r2:Status</th>\n",
       "      <th>CrPs* -N r1:PED</th>\n",
       "      <th>CrPs* -N r1:COL</th>\n",
       "      <th>CrPs* -N r1:Status</th>\n",
       "      <th>Cr*Ps -P r2:PED</th>\n",
       "      <th>Cr*Ps -P r2:COL</th>\n",
       "      <th>Cr*Ps -P r2:Status</th>\n",
       "      <th>ID</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>0, 0</td>\n",
       "      <td>564, 914</td>\n",
       "      <td>not mobile</td>\n",
       "      <td>299, 164</td>\n",
       "      <td>0, 0</td>\n",
       "      <td>not mobile</td>\n",
       "      <td>937, 1178</td>\n",
       "      <td>4, 0</td>\n",
       "      <td>mobile</td>\n",
       "      <td>0, 0</td>\n",
       "      <td>...</td>\n",
       "      <td>470, 597</td>\n",
       "      <td>0, 0</td>\n",
       "      <td>not mobile</td>\n",
       "      <td>456, 276</td>\n",
       "      <td>0, 0</td>\n",
       "      <td>not mobile</td>\n",
       "      <td>0, 0</td>\n",
       "      <td>268, 232</td>\n",
       "      <td>not mobile</td>\n",
       "      <td>AT1G50920.1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>0, 0, 0, 0, 0</td>\n",
       "      <td>95, 147, 114, 108, 32</td>\n",
       "      <td>not mobile</td>\n",
       "      <td>52, 55, 21, 35, 8</td>\n",
       "      <td>0, 0, 0, 0, 0</td>\n",
       "      <td>not mobile</td>\n",
       "      <td>87, 134, 135, 123, 54</td>\n",
       "      <td>0, 0, 0, 0, 0</td>\n",
       "      <td>not mobile</td>\n",
       "      <td>0, 0, 0, 0, 0</td>\n",
       "      <td>...</td>\n",
       "      <td>30, 35, 28, 79, 22</td>\n",
       "      <td>0, 0, 0, 0, 0</td>\n",
       "      <td>not mobile</td>\n",
       "      <td>50, 70, 21, 61, 7</td>\n",
       "      <td>0, 0, 0, 0, 0</td>\n",
       "      <td>not mobile</td>\n",
       "      <td>0, 0, 0, 0, 0</td>\n",
       "      <td>63, 73, 38, 76, 18</td>\n",
       "      <td>not mobile</td>\n",
       "      <td>AT1G73440.1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>0, 0, 0</td>\n",
       "      <td>40, 32, 6</td>\n",
       "      <td>not mobile</td>\n",
       "      <td>21, 10, NA</td>\n",
       "      <td>0, 0, NA</td>\n",
       "      <td>not mobile</td>\n",
       "      <td>131, 125, 30</td>\n",
       "      <td>0, 0, 0</td>\n",
       "      <td>not mobile</td>\n",
       "      <td>0, 0, 0</td>\n",
       "      <td>...</td>\n",
       "      <td>27, 35, 15</td>\n",
       "      <td>0, 0, 0</td>\n",
       "      <td>not mobile</td>\n",
       "      <td>33, 17, NA</td>\n",
       "      <td>0, 0, NA</td>\n",
       "      <td>not mobile</td>\n",
       "      <td>0, 0, NA</td>\n",
       "      <td>20, 8, NA</td>\n",
       "      <td>not mobile</td>\n",
       "      <td>AT1G51380.1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>0</td>\n",
       "      <td>820</td>\n",
       "      <td>not mobile</td>\n",
       "      <td>291</td>\n",
       "      <td>0</td>\n",
       "      <td>not mobile</td>\n",
       "      <td>1187</td>\n",
       "      <td>0</td>\n",
       "      <td>not mobile</td>\n",
       "      <td>0</td>\n",
       "      <td>...</td>\n",
       "      <td>222</td>\n",
       "      <td>0</td>\n",
       "      <td>not mobile</td>\n",
       "      <td>286</td>\n",
       "      <td>0</td>\n",
       "      <td>not mobile</td>\n",
       "      <td>0</td>\n",
       "      <td>192</td>\n",
       "      <td>not mobile</td>\n",
       "      <td>AT1G10950.1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0...</td>\n",
       "      <td>37, 131, 102, 80, 32, 31, 24, 18, 19, 20, 23, ...</td>\n",
       "      <td>not mobile</td>\n",
       "      <td>30, 40, 36, 36, 18, 18, NA, NA, NA, NA, NA, 4,...</td>\n",
       "      <td>0, 0, 0, 0, 0, 0, NA, NA, NA, NA, NA, 0, 0, 0,...</td>\n",
       "      <td>not mobile</td>\n",
       "      <td>86, 103, 93, 84, 52, 52, 18, NA, NA, NA, 5, 3,...</td>\n",
       "      <td>0, 0, 0, 0, 0, 0, 0, NA, NA, NA, 0, 0, 0, 0, 0...</td>\n",
       "      <td>not mobile</td>\n",
       "      <td>0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0...</td>\n",
       "      <td>...</td>\n",
       "      <td>25, 33, 30, 28, 21, 13, 11, 9, 9, 7, 14, 12, 1...</td>\n",
       "      <td>0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0...</td>\n",
       "      <td>not mobile</td>\n",
       "      <td>38, 42, 45, 43, 29, 26, 22, 10, 8, 7, 11, 12, ...</td>\n",
       "      <td>0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0...</td>\n",
       "      <td>not mobile</td>\n",
       "      <td>0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0...</td>\n",
       "      <td>29, 79, 64, 72, 60, 61, 49, 30, 34, 32, 26, 22...</td>\n",
       "      <td>not mobile</td>\n",
       "      <td>AT1G31870.1</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5 rows Ã— 88 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                              PrCs* FN_stemUpper:PED  \\\n",
       "0                                               0, 0   \n",
       "1                                      0, 0, 0, 0, 0   \n",
       "2                                            0, 0, 0   \n",
       "3                                                  0   \n",
       "4  0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0...   \n",
       "\n",
       "                              PrCs* FN_stemUpper:COL  \\\n",
       "0                                           564, 914   \n",
       "1                              95, 147, 114, 108, 32   \n",
       "2                                          40, 32, 6   \n",
       "3                                                820   \n",
       "4  37, 131, 102, 80, 32, 31, 24, 18, 19, 20, 23, ...   \n",
       "\n",
       "  PrCs* FN_stemUpper:Status  \\\n",
       "0                not mobile   \n",
       "1                not mobile   \n",
       "2                not mobile   \n",
       "3                not mobile   \n",
       "4                not mobile   \n",
       "\n",
       "                                     CrPs* -P r1:PED  \\\n",
       "0                                           299, 164   \n",
       "1                                  52, 55, 21, 35, 8   \n",
       "2                                         21, 10, NA   \n",
       "3                                                291   \n",
       "4  30, 40, 36, 36, 18, 18, NA, NA, NA, NA, NA, 4,...   \n",
       "\n",
       "                                     CrPs* -P r1:COL CrPs* -P r1:Status  \\\n",
       "0                                               0, 0         not mobile   \n",
       "1                                      0, 0, 0, 0, 0         not mobile   \n",
       "2                                           0, 0, NA         not mobile   \n",
       "3                                                  0         not mobile   \n",
       "4  0, 0, 0, 0, 0, 0, NA, NA, NA, NA, NA, 0, 0, 0,...         not mobile   \n",
       "\n",
       "                                   Pr*Cs FN_root:PED  \\\n",
       "0                                          937, 1178   \n",
       "1                              87, 134, 135, 123, 54   \n",
       "2                                       131, 125, 30   \n",
       "3                                               1187   \n",
       "4  86, 103, 93, 84, 52, 52, 18, NA, NA, NA, 5, 3,...   \n",
       "\n",
       "                                   Pr*Cs FN_root:COL Pr*Cs FN_root:Status  \\\n",
       "0                                               4, 0               mobile   \n",
       "1                                      0, 0, 0, 0, 0           not mobile   \n",
       "2                                            0, 0, 0           not mobile   \n",
       "3                                                  0           not mobile   \n",
       "4  0, 0, 0, 0, 0, 0, 0, NA, NA, NA, 0, 0, 0, 0, 0...           not mobile   \n",
       "\n",
       "                                     Cr*Ps -N r2:PED  ...  \\\n",
       "0                                               0, 0  ...   \n",
       "1                                      0, 0, 0, 0, 0  ...   \n",
       "2                                            0, 0, 0  ...   \n",
       "3                                                  0  ...   \n",
       "4  0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0...  ...   \n",
       "\n",
       "                                        Pr*Cs r2:PED  \\\n",
       "0                                           470, 597   \n",
       "1                                 30, 35, 28, 79, 22   \n",
       "2                                         27, 35, 15   \n",
       "3                                                222   \n",
       "4  25, 33, 30, 28, 21, 13, 11, 9, 9, 7, 14, 12, 1...   \n",
       "\n",
       "                                        Pr*Cs r2:COL Pr*Cs r2:Status  \\\n",
       "0                                               0, 0      not mobile   \n",
       "1                                      0, 0, 0, 0, 0      not mobile   \n",
       "2                                            0, 0, 0      not mobile   \n",
       "3                                                  0      not mobile   \n",
       "4  0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0...      not mobile   \n",
       "\n",
       "                                     CrPs* -N r1:PED  \\\n",
       "0                                           456, 276   \n",
       "1                                  50, 70, 21, 61, 7   \n",
       "2                                         33, 17, NA   \n",
       "3                                                286   \n",
       "4  38, 42, 45, 43, 29, 26, 22, 10, 8, 7, 11, 12, ...   \n",
       "\n",
       "                                     CrPs* -N r1:COL CrPs* -N r1:Status  \\\n",
       "0                                               0, 0         not mobile   \n",
       "1                                      0, 0, 0, 0, 0         not mobile   \n",
       "2                                           0, 0, NA         not mobile   \n",
       "3                                                  0         not mobile   \n",
       "4  0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0...         not mobile   \n",
       "\n",
       "                                     Cr*Ps -P r2:PED  \\\n",
       "0                                               0, 0   \n",
       "1                                      0, 0, 0, 0, 0   \n",
       "2                                           0, 0, NA   \n",
       "3                                                  0   \n",
       "4  0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0...   \n",
       "\n",
       "                                     Cr*Ps -P r2:COL Cr*Ps -P r2:Status  \\\n",
       "0                                           268, 232         not mobile   \n",
       "1                                 63, 73, 38, 76, 18         not mobile   \n",
       "2                                          20, 8, NA         not mobile   \n",
       "3                                                192         not mobile   \n",
       "4  29, 79, 64, 72, 60, 61, 49, 30, 34, 32, 26, 22...         not mobile   \n",
       "\n",
       "            ID  \n",
       "0  AT1G50920.1  \n",
       "1  AT1G73440.1  \n",
       "2  AT1G51380.1  \n",
       "3  AT1G10950.1  \n",
       "4  AT1G31870.1  \n",
       "\n",
       "[5 rows x 88 columns]"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "## Data taken from: \"Endogenous Arabidopsis messenger RNAs transported to distant tissues\" Thieme et al. 2015\n",
    "## Code to run for table 1 - runs in < 1 sec.\n",
    "\n",
    "exps = [col for col in df.columns if \"*\" in col]\n",
    "\n",
    "## The columns we are using: COL base counts, PED base counts\n",
    "exps_split = [exp.split(\":\")[0] for exp in exps]\n",
    "exps = set(exps_split)\n",
    "df_thieme_list = []\n",
    "\n",
    "for exp in exps:\n",
    "    col = exp + \":COL\"\n",
    "    ped = exp + \":PED\"\n",
    "    mobile = exp + \":Status\"\n",
    "    \n",
    "    col_base_counts = df.filter(like=col)\n",
    "    ped_base_counts = df.filter(like=ped)\n",
    "    mobile_status = df.filter(like=mobile)\n",
    "\n",
    "    # Rename columns\n",
    "    col_base_counts.columns = [col]\n",
    "    ped_base_counts.columns = [ped]\n",
    "    mobile_status.columns = [mobile]\n",
    "    \n",
    "    df_thieme_list.append(pd.concat([ped_base_counts, col_base_counts, mobile_status], axis=1))\n",
    "\n",
    "# Convert df_thieme_list to a DataFrame\n",
    "df_thieme = pd.concat(df_thieme_list, axis=1)\n",
    "df_thieme[\"ID\"] = df.filter(like=\"Transcript:\")\n",
    "\n",
    "display(df_thieme.head())"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [],
   "source": [
    "## Run for 2 different error rates: 1 in 1000 and 1 in 10000. This cell runs in ~ 23 mins.\n",
    "output1000_list = []\n",
    "output10000_list = []\n",
    "\n",
    "for exp in exps:\n",
    "\n",
    "    output1000 = df_thieme.filter(like=exp).copy()\n",
    "    output10000 = df_thieme.filter(like=exp).copy()\n",
    "\n",
    "    ## Rename columns that include the word \"Status\" as \"Status\"\n",
    "    output1000.rename(columns={exp + \":Status\": \"Status\"}, inplace=True)\n",
    "    output10000.rename(columns={exp + \":Status\": \"Status\"}, inplace=True)\n",
    "    \n",
    "    ## Get whether the sample is from col or ped\n",
    "    if \"Ps*\" in exp or \"Pr*\" in exp:\n",
    "         output1000.rename(columns={exp + \":PED\": \"endo\", exp + \":COL\": \"distal\"}, inplace=True)\n",
    "         output10000.rename(columns={exp + \":PED\": \"endo\", exp + \":COL\": \"distal\"}, inplace=True)\n",
    "    elif \"Cs*\" in exp or \"Cr*\" in exp:\n",
    "         output1000.rename(columns={exp + \":COL\": \"endo\", exp + \":PED\": \"distal\"}, inplace=True)\n",
    "         output10000.rename(columns={exp + \":COL\": \"endo\", exp + \":PED\": \"distal\"}, inplace=True)\n",
    "    else:\n",
    "        print(\"Error\")\n",
    "\n",
    "    output1000[\"log10BF\"] = output1000.parallel_apply(calculate_bf, args=(1000,), axis=1)\n",
    "    output10000[\"log10BF\"] = output10000.parallel_apply(calculate_bf, args=(10000,), axis=1)\n",
    "    output1000[\"exp\"] = exp\n",
    "    output10000[\"exp\"] = exp\n",
    "    output1000.dropna(inplace=True)\n",
    "    output10000.dropna(inplace=True)\n",
    "\n",
    "    ## Uncomment the below to see the results for each experiment\n",
    "\n",
    "    #print(\"Experiment: \", exp)\n",
    "\n",
    "    #print(\"Error = 1 in 1000\")\n",
    "    #print(\"Total Thieme find to be mobile:\")\n",
    "    #print(len(output1000[output1000[\"Status\"] == \"mobile\"]))\n",
    "    #print(\"Total BF find to be mobile:\")\n",
    "    #print(len(output1000[output1000[\"log10BF\"] >= 1]))\n",
    "\n",
    "    #print(\"Error = 1 in 10000\")\n",
    "    #print(len(output10000[output10000[\"Status\"] == \"mobile\"]))\n",
    "    #print(\"Total Thieme find to be mobile:\")\n",
    "    #print(\"Total BF find to be mobile:\")\n",
    "    #print(len(output10000[output10000[\"log10BF\"] >= 1]))\n",
    "\n",
    "    output1000_list.append(output1000)\n",
    "    output10000_list.append(output10000)\n",
    "    "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>distal</th>\n",
       "      <th>endo</th>\n",
       "      <th>Status</th>\n",
       "      <th>log10BF</th>\n",
       "      <th>exp</th>\n",
       "      <th>ID</th>\n",
       "      <th>Transcript</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>0, 0</td>\n",
       "      <td>564, 914</td>\n",
       "      <td>not mobile</td>\n",
       "      <td>-4</td>\n",
       "      <td>PrCs* FN_stemUpper</td>\n",
       "      <td>AT1G50920.1</td>\n",
       "      <td>AT1G50920</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>0, 0, 0, 0, 0</td>\n",
       "      <td>95, 147, 114, 108, 32</td>\n",
       "      <td>not mobile</td>\n",
       "      <td>-10</td>\n",
       "      <td>PrCs* FN_stemUpper</td>\n",
       "      <td>AT1G73440.1</td>\n",
       "      <td>AT1G73440</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>0, 0, 0</td>\n",
       "      <td>40, 32, 6</td>\n",
       "      <td>not mobile</td>\n",
       "      <td>-6</td>\n",
       "      <td>PrCs* FN_stemUpper</td>\n",
       "      <td>AT1G51380.1</td>\n",
       "      <td>AT1G51380</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>0</td>\n",
       "      <td>820</td>\n",
       "      <td>not mobile</td>\n",
       "      <td>-2</td>\n",
       "      <td>PrCs* FN_stemUpper</td>\n",
       "      <td>AT1G10950.1</td>\n",
       "      <td>AT1G10950</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0...</td>\n",
       "      <td>37, 131, 102, 80, 32, 31, 24, 18, 19, 20, 23, ...</td>\n",
       "      <td>not mobile</td>\n",
       "      <td>-70</td>\n",
       "      <td>PrCs* FN_stemUpper</td>\n",
       "      <td>AT1G31870.1</td>\n",
       "      <td>AT1G31870</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                                              distal  \\\n",
       "0                                               0, 0   \n",
       "1                                      0, 0, 0, 0, 0   \n",
       "2                                            0, 0, 0   \n",
       "3                                                  0   \n",
       "4  0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0...   \n",
       "\n",
       "                                                endo      Status log10BF  \\\n",
       "0                                           564, 914  not mobile      -4   \n",
       "1                              95, 147, 114, 108, 32  not mobile     -10   \n",
       "2                                          40, 32, 6  not mobile      -6   \n",
       "3                                                820  not mobile      -2   \n",
       "4  37, 131, 102, 80, 32, 31, 24, 18, 19, 20, 23, ...  not mobile     -70   \n",
       "\n",
       "                  exp           ID Transcript  \n",
       "0  PrCs* FN_stemUpper  AT1G50920.1  AT1G50920  \n",
       "1  PrCs* FN_stemUpper  AT1G73440.1  AT1G73440  \n",
       "2  PrCs* FN_stemUpper  AT1G51380.1  AT1G51380  \n",
       "3  PrCs* FN_stemUpper  AT1G10950.1  AT1G10950  \n",
       "4  PrCs* FN_stemUpper  AT1G31870.1  AT1G31870  "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "## Create the final dataframes and add in the transcript information\n",
    "\n",
    "df_output1000 = pd.concat(output1000_list)\n",
    "df_output10000 = pd.concat(output10000_list)\n",
    "\n",
    "df_output1000[\"ID\"] = df_thieme[\"ID\"]\n",
    "df_output10000[\"ID\"] = df_thieme[\"ID\"]\n",
    "\n",
    "df_output1000[\"Transcript\"] = df_output1000[\"ID\"].apply(lambda x: x.split(\".\")[0])\n",
    "df_output10000[\"Transcript\"] = df_output10000[\"ID\"].apply(lambda x: x.split(\".\")[0])\n",
    "\n",
    "display(df_output1000.head())"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [],
   "source": [
    "## Save the processed data\n",
    "df_output1000.to_csv(\"../Data/thieme_bf_1000.csv\")\n",
    "df_output10000.to_csv(\"../Data/thieme_bf_10000.csv\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Thieme mobile: 2006\n",
      "logBF>=1: 551\n",
      "Thieme mobile: 2006\n",
      "logBF >= 1: 920\n"
     ]
    }
   ],
   "source": [
    "## For each error rates, print the number of transcripts with a log10BF >= 1, and the amount found mobile in Thieme et al. 2015\n",
    "\n",
    "## Import the data\n",
    "df_output1000 = pd.read_csv(\"../Data/thieme_bf_1000.csv\")\n",
    "df_output10000 = pd.read_csv(\"../Data/thieme_bf_10000.csv\")\n",
    "\n",
    "## 1 in 1000\n",
    "thieme_mobile = set(df_output1000[df_output1000[\"Status\"]==\"mobile\"][\"Transcript\"].to_list())\n",
    "bf_mobile = set(df_output1000[df_output1000[\"log10BF\"]>=1][\"Transcript\"].to_list())\n",
    "\n",
    "print(f\"Thieme mobile: {len(thieme_mobile)}\")\n",
    "print(f\"logBF>=1: {len(bf_mobile)}\")\n",
    "\n",
    "## 1 in 10000\n",
    "thieme_mobile = set(df_output10000[df_output10000[\"Status\"]==\"mobile\"][\"Transcript\"].to_list())\n",
    "bf_mobile = set(df_output10000[df_output10000[\"log10BF\"]>=1][\"Transcript\"].to_list())\n",
    "\n",
    "print(f\"Thieme mobile: {len(thieme_mobile)}\")\n",
    "print(f\"logBF >= 1: {len(bf_mobile)}\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/var/folders/t5/q1tt12hd2sg29kw894zcfbtm000b_5/T/ipykernel_55375/3631832613.py:14: FutureWarning: Downcasting behavior in `replace` is deprecated and will be removed in a future version. To retain the old behavior, explicitly call `result.infer_objects(copy=False)`. To opt-in to the future behavior, set `pd.set_option('future.no_silent_downcasting', True)`\n",
      "  df_vgir.replace(\"-\", 0, inplace=True)\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>V. girdiana scion-1:RPKM</th>\n",
       "      <th>endo</th>\n",
       "      <th>distal</th>\n",
       "      <th>V. girdiana scion-1:No. unique reads</th>\n",
       "      <th>SNP</th>\n",
       "      <th>N</th>\n",
       "      <th>n</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>11.20</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>GSVIVG01000021001</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>4.90</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>GSVIVG01000034001</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>8.98</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>GSVIVG01000035001</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>21.60</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>GSVIVG01000037001</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>14.34</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>GSVIVG01000040001</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "   V. girdiana scion-1:RPKM  endo  distal  \\\n",
       "0                     11.20     0       0   \n",
       "1                      4.90     0       0   \n",
       "2                      8.98     0       0   \n",
       "3                     21.60     0       0   \n",
       "4                     14.34     0       0   \n",
       "\n",
       "   V. girdiana scion-1:No. unique reads                SNP  N  n  \n",
       "0                                     0  GSVIVG01000021001  0  0  \n",
       "1                                     0  GSVIVG01000034001  0  0  \n",
       "2                                     0  GSVIVG01000035001  0  0  \n",
       "3                                     0  GSVIVG01000037001  0  0  \n",
       "4                                     0  GSVIVG01000040001  0  0  "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "## Same process as above, but for the grapevine data\n",
    "\n",
    "df = pd.read_excel(\"../Data/Vitis_vinifera_mobile_mRNA.xlsx\", sheet_name=\"Dataset S2\", skiprows=1, header=[0, 1])\n",
    "\n",
    "# Merge the two levels of the header into a single level with a more meaningful separator\n",
    "df.columns = [':'.join(col).strip() for col in df.columns.values]\n",
    "\n",
    "## Get the columns relating to V. girdiana rootstock\n",
    "df_vgir = df.filter(like=\"V. girdiana scion-1\").copy()\n",
    "\n",
    "df_vgir[\"SNP\"] = df[\"Gene ID:Unnamed: 0_level_1\"]\n",
    "\n",
    "## Replace the dashes with 0s\n",
    "df_vgir.replace(\"-\", 0, inplace=True)\n",
    "\n",
    "df_vgir.rename(columns={\"V. girdiana scion-1:No. source reads\": \"endo\", \"V. girdiana scion-1:No. reads\": \"distal\"}, inplace=True)\n",
    "\n",
    "df_vgir[\"N\"] = df_vgir[\"endo\"] + df_vgir[\"distal\"]\n",
    "df_vgir[\"n\"] = df_vgir[\"distal\"]\n",
    "\n",
    "display(df_vgir.head())\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Warning: 2203 Zero values were removed.\n",
      "INFO: Pandarallel will run on 10 workers.\n",
      "INFO: Pandarallel will use standard multiprocessing data transfer (pipe) to transfer data between the main process and workers.\n"
     ]
    },
    {
     "data": {
      "application/vnd.jupyter.widget-view+json": {
       "model_id": "5483b0abe1a84f00852ba830b62d86d9",
       "version_major": 2,
       "version_minor": 0
      },
      "text/plain": [
       "VBox(children=(HBox(children=(IntProgress(value=0, description='0.00%', max=113), Label(value='0 / 113'))), HBâ€¦"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Numbers of mobile found in paper:\n",
      "1130\n",
      "No. consistent with error of 0.1%:\n",
      "945\n",
      "Warning: 2203 Zero values were removed.\n",
      "INFO: Pandarallel will run on 10 workers.\n",
      "INFO: Pandarallel will use standard multiprocessing data transfer (pipe) to transfer data between the main process and workers.\n"
     ]
    },
    {
     "data": {
      "application/vnd.jupyter.widget-view+json": {
       "model_id": "e8e9697d5775443eba7a52be00b041f0",
       "version_major": 2,
       "version_minor": 0
      },
      "text/plain": [
       "VBox(children=(HBox(children=(IntProgress(value=0, description='0.00%', max=113), Label(value='0 / 113'))), HBâ€¦"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Numbers of mobile found in paper:\n",
      "1130\n",
      "No. consistent with error of 0.01%:\n",
      "384\n"
     ]
    }
   ],
   "source": [
    "## Calculate the errors rates for 1 in 1000 and 1 in 10000 - runs in less than 10 secs.\n",
    "error1 = 1000\n",
    "error2 = 10000\n",
    "\n",
    "## Run for 1 in 1000\n",
    "\n",
    "df_vgir[\"Nh1\"] = error1\n",
    "df_vgir[\"nh1\"] = 1\n",
    "df_vgir[\"Nh2\"] = error1\n",
    "df_vgir[\"nh2\"] = 1\n",
    "\n",
    "## Convert N and n to integers\n",
    "df_vgir[\"N\"] = df_vgir[\"N\"].astype(int)\n",
    "df_vgir[\"n\"] = df_vgir[\"n\"].astype(int)\n",
    "\n",
    "df_vgir1000 = baymob.run_bayes(df_vgir,10)\n",
    "print(\"Numbers of mobile found in paper:\")\n",
    "print(len(df_vgir1000))\n",
    "print(\"No. consistent with error of 0.1%:\")\n",
    "print(len(df_vgir1000) - len(df_vgir1000[df_vgir1000[\"log10BF\"]>=1]))\n",
    "\n",
    "## Run for 1 in 10000\n",
    "\n",
    "df_vgir[\"Nh1\"] = error2\n",
    "df_vgir[\"nh1\"] = 1\n",
    "df_vgir[\"Nh2\"] = error2\n",
    "df_vgir[\"nh2\"] = 1\n",
    "\n",
    "## Convert N and n to integers\n",
    "df_vgir[\"N\"] = df_vgir[\"N\"].astype(int)\n",
    "df_vgir[\"n\"] = df_vgir[\"n\"].astype(int)\n",
    "\n",
    "df_vgir10000 = baymob.run_bayes(df_vgir,10)\n",
    "print(\"Numbers of mobile found in paper:\")\n",
    "print(len(df_vgir10000))\n",
    "print(\"No. consistent with error of 0.01%:\")\n",
    "print(len(df_vgir10000) - len(df_vgir10000[df_vgir10000[\"log10BF\"]>=1]))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "base",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.12.7"
  },
  "orig_nbformat": 4
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
