{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "## Reanalyse the Thieme data based on the Excel spreadsheet.\n",
    "## Choose an error value\n",
    "\n",
    "## Question: do we see any consistency of function in the \"false positives\" between Thieme and grapevine?\n",
    "\n",
    "import pandas as pd\n",
    "\n",
    "df = pd.read_excel(\"thieme_original_data.xlsx\", sheet_name=\"SNP in all transcripts\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "import baymobil as baymob\n",
    "\n",
    "exps = [col for col in df.columns if \"*\" in col]\n",
    "output = []\n",
    "## Calculate the BF for each entry in each experiment, summing over the SNPs\n",
    "\n",
    "def calculate_bf(row):\n",
    "    bf_list = []\n",
    "    ## Separate into ints\n",
    "    ## Check if there are multiple values\n",
    "    if \",\" in str(row.N):\n",
    "        N_list = row.N.split(\",\")\n",
    "        n_list = row.n.split(\",\")\n",
    "        ## Remove the poxy \"NA\" values (why? why? why?)\n",
    "        N_list = [int(n) for n in N_list if \"NA\" not in n]\n",
    "        n_list = [int(n) for n in n_list if \"NA\" not in n]\n",
    "    else:\n",
    "        try:\n",
    "            N_list = [int(row.N)]\n",
    "            n_list = [int(row.n)]\n",
    "        except:\n",
    "            print(\"NAN\")\n",
    "            return \"NA\"\n",
    "    \n",
    "    for N, n in zip(N_list, n_list):\n",
    "        bf = baymob.fasterpostN2(500,1,500,1,int(N) + int(n), int(n), 10)\n",
    "        ## Cap the BF values\n",
    "        if bf[2] > 2: bf[2] = 2\n",
    "        if bf[2] <-2: bf[2] = -2\n",
    "        bf_list.append(bf[2])\n",
    "    return sum(bf_list)\n",
    "\n",
    "for exp in exps:\n",
    "    ind = df.columns.get_loc(exp)\n",
    "    df_subset = df.iloc[:,ind:ind +2]\n",
    "    df_subset[\"Transcript\"] = df[\"Transcript\"]\n",
    "    df_subset[\"Mobile\"] = df.iloc[:,ind+7]\n",
    "    ## Get rid of NaNs (but not NAs, because we hate you.)\n",
    "    df_subset.dropna(inplace=True)\n",
    "    df_subset = df_subset[1:] #take the data less the header row\n",
    "    ## Get the sampling location\n",
    "    star_index = exp.index(\"*\")\n",
    "    sample = exp[star_index-2:star_index]\n",
    "    if sample[0] == \"C\":\n",
    "        type = \"Col\"\n",
    "    if sample[0] == \"P\":\n",
    "        type = \"Ped\"\n",
    "    if sample[1] == \"r\":\n",
    "        tissue = \"root\"\n",
    "    if sample[1] == \"s\":\n",
    "        tissue = \"shoot\"\n",
    "    if type == \"Col\":\n",
    "        df_subset.columns = [\"N\",\"n\", \"Transcript\",\"Mobile\"]\n",
    "    if type == \"Ped\":\n",
    "        df_subset.columns = [\"n\",\"N\", \"Transcript\",\"Mobile\"]\n",
    "    df_subset[\"log10BF\"] = df_subset.apply(lambda x: calculate_bf(x), axis= 1)\n",
    "    output.append(df_subset)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "output_exp = []\n",
    "for df, exp in zip(output, exps):\n",
    "    print(exp)\n",
    "    print(\"Total Thieme find to be mobile:\")\n",
    "    print(len(df[df[\"Mobile\"] == \"mobile\"]))\n",
    "    print(\"Total BF find to be mobile:\")\n",
    "    print(len(df[df[\"log10BF\"] >=1]))\n",
    "    df[\"exp\"] = exp\n",
    "    output_exp.append(df)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "df_output = pd.concat(output_exp)\n",
    "df_output[\"Gene\"] = df_output[\"Transcript\"].apply(lambda x: x.split(\".\")[0])\n",
    "thieme_mobile = set(df_output[df_output[\"Mobile\"]==\"mobile\"][\"Gene\"].to_list())\n",
    "bf_mobile = set(df_output[df_output[\"log10BF\"]>=1][\"Gene\"].to_list())\n",
    "\n",
    "print(len(thieme_mobile))\n",
    "print(len(bf_mobile))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "df_output[df_output[\"Transcript\"]==\"AT1G01010.1\"]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "display(df_output[(df_output[\"Mobile\"]==\"mobile\")])\n",
    "\n",
    "df_output.to_csv(\"thieme_1_500.csv\", index=None)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "pd.set_option(\"display.max_rows\",100)\n",
    "\n",
    "df_sorted = df_output.sort_values(by=\"log10BF\", ascending = False)\n",
    "\n",
    "display(df_sorted[(df_sorted[\"Mobile\"] == \"mobile\")].head(99))\n",
    "display(df_sorted[(df_sorted[\"Mobile\"] == \"mobile\")].head(99)[[\"Transcript\",\"exp\",\"Total reads\",\"Mobile reads\",\"No. of SNPs\",\"log10BF\"]])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "## Check the homograft\n",
    "\n",
    "df_snp = pd.read_csv(\"raw_data/thieme/snp_ref.csv\")\n",
    "\n",
    "snps = df_snp[df_snp[\"transcripts\"]==\"AT3G45140\"][\"SNP\"].to_list()\n",
    "\n",
    "df_hom1 = pd.read_csv(\"raw_data/thieme/homfiles/C-C-shoot-FN.txt\", delimiter=\"\\t\")\n",
    "df_hom1[df_hom1[\"SNP\"].isin(snps)]\n",
    "\n",
    "df_hom2 = pd.read_csv(\"raw_data/thieme/homfiles/P-P-shoot-FN.txt\", delimiter=\"\\t\")\n",
    "df_hom2[df_hom2[\"SNP\"].isin(snps)]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "## Sum the number of endogenous : distal reads per transcript\n",
    "\n",
    "def sumN(row):\n",
    "    if \",\" in str(row):\n",
    "        N_list = row.split(\",\")\n",
    "        ## Remove the poxy \"NA\" values (why? why? why?)\n",
    "        N_list = [int(n) for n in N_list if \"NA\" not in n]\n",
    "    else:\n",
    "        try:\n",
    "            N_list = [int(row)]\n",
    "        except:\n",
    "            print(\"NAN\")\n",
    "            return \"NA\"\n",
    "    return sum(N_list)\n",
    "\n",
    "def countN(row):\n",
    "    if \",\" in str(row):\n",
    "        N_list = row.split(\",\")\n",
    "        ## Remove the poxy \"NA\" values (why? why? why?)\n",
    "        N_list = [int(n) for n in N_list if \"NA\" not in n]\n",
    "    else:\n",
    "        try:\n",
    "            N_list = [int(row)]\n",
    "        except:\n",
    "            print(\"NAN\")\n",
    "            return \"NA\"\n",
    "    return len(N_list)\n",
    "    \n",
    "df_output[\"sum_N\"] = df_output[\"N\"].apply(lambda x: sumN(x))\n",
    "df_output[\"sum_n\"] = df_output[\"n\"].apply(lambda x: sumN(x))\n",
    "\n",
    "df_output[\"No. of SNPs\"] = df_output[\"N\"].apply(lambda x: countN(x))\n",
    "df_output"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "ans = df_output[(df_output[\"Mobile\"] == \"mobile\") & ((df_output[\"sum_n\"] / (df_output[\"sum_N\"] + df_output[\"sum_n\"])) < (1 / 500))]\n",
    "\n",
    "\n",
    "overs = df_output[(df_output[\"Mobile\"] == \"mobile\") & ((df_output[\"sum_n\"] / (df_output[\"sum_N\"] + df_output[\"sum_n\"])) > (1 / 500))]\n",
    "\n",
    "ans = ans[~ans[\"Gene\"].isin(overs[\"Gene\"].to_list())]\n",
    "len(set(ans[\"Gene\"].to_list()))\n",
    "## Need the ones that only appear once - or all occurences have the low ratio. Why is nothing simple??/\n",
    "## Select those with reads above 1: 500, and take those out of the list"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 186,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Transcript</th>\n",
       "      <th>snp_list</th>\n",
       "      <th>SNP</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>AT1G50920.1</td>\n",
       "      <td>220</td>\n",
       "      <td>Chr1_18870656</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>AT1G50920.1</td>\n",
       "      <td>1351</td>\n",
       "      <td>Chr1_18871787</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>AT1G51380.1</td>\n",
       "      <td>209</td>\n",
       "      <td>Chr1_19048090</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>AT1G51380.1</td>\n",
       "      <td>987</td>\n",
       "      <td>Chr1_19049605</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>AT1G51380.1</td>\n",
       "      <td>1376</td>\n",
       "      <td>Chr1_19050086</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>54648</th>\n",
       "      <td>AT5G25560.4</td>\n",
       "      <td>1361</td>\n",
       "      <td>Chr5_8899341</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>54649</th>\n",
       "      <td>AT5G25560.4</td>\n",
       "      <td>1418</td>\n",
       "      <td>Chr5_8899398</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>54650</th>\n",
       "      <td>AT5G25560.4</td>\n",
       "      <td>1465</td>\n",
       "      <td>Chr5_8899445</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>54651</th>\n",
       "      <td>AT5G25560.4</td>\n",
       "      <td>1466</td>\n",
       "      <td>Chr5_8899446</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>54652</th>\n",
       "      <td>AT5G25560.4</td>\n",
       "      <td>1543</td>\n",
       "      <td>Chr5_8899523</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>54653 rows × 3 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "        Transcript  snp_list            SNP\n",
       "0      AT1G50920.1       220  Chr1_18870656\n",
       "1      AT1G50920.1      1351  Chr1_18871787\n",
       "2      AT1G51380.1       209  Chr1_19048090\n",
       "3      AT1G51380.1       987  Chr1_19049605\n",
       "4      AT1G51380.1      1376  Chr1_19050086\n",
       "...            ...       ...            ...\n",
       "54648  AT5G25560.4      1361   Chr5_8899341\n",
       "54649  AT5G25560.4      1418   Chr5_8899398\n",
       "54650  AT5G25560.4      1465   Chr5_8899445\n",
       "54651  AT5G25560.4      1466   Chr5_8899446\n",
       "54652  AT5G25560.4      1543   Chr5_8899523\n",
       "\n",
       "[54653 rows x 3 columns]"
      ]
     },
     "execution_count": 186,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "## Load in the new snp list\n",
    "\n",
    "df_snp = pd.read_csv(\"new_thieme_og_snps.csv\", index_col=None)\n",
    "df_snp"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/Users/tomkinsm/miniconda3/lib/python3.10/site-packages/openpyxl/worksheet/_reader.py:329: UserWarning: Unknown extension is not supported and will be removed\n",
      "  warn(msg)\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead tr th {\n",
       "        text-align: left;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr>\n",
       "      <th></th>\n",
       "      <th colspan=\"2\" halign=\"left\">Cr*Ps r1</th>\n",
       "      <th colspan=\"2\" halign=\"left\">Cr*Ps r2</th>\n",
       "      <th colspan=\"2\" halign=\"left\">Cr*Ps r3</th>\n",
       "      <th colspan=\"2\" halign=\"left\">PrCs* r1</th>\n",
       "      <th colspan=\"2\" halign=\"left\">PrCs* r2</th>\n",
       "      <th>...</th>\n",
       "      <th colspan=\"2\" halign=\"left\">CrPs* -N r3</th>\n",
       "      <th colspan=\"2\" halign=\"left\">CrPs* -P r1</th>\n",
       "      <th colspan=\"2\" halign=\"left\">CrPs* -P r2</th>\n",
       "      <th colspan=\"2\" halign=\"left\">CrPs* -P r3</th>\n",
       "      <th colspan=\"2\" halign=\"left\">Pr*Cs FN_root</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th></th>\n",
       "      <th>COL base counts per informative SNP site (NA in cases of no read confirming either of two alleles)</th>\n",
       "      <th>PED base counts per informative SNP site</th>\n",
       "      <th>COL base counts per informative SNP site</th>\n",
       "      <th>PED base counts per informative SNP site</th>\n",
       "      <th>COL base counts per informative SNP site</th>\n",
       "      <th>PED base counts per informative SNP site</th>\n",
       "      <th>COL base counts per informative SNP site</th>\n",
       "      <th>PED base counts per informative SNP site</th>\n",
       "      <th>COL base counts per informative SNP site</th>\n",
       "      <th>PED base counts per informative SNP site</th>\n",
       "      <th>...</th>\n",
       "      <th>COL base counts per informative SNP site</th>\n",
       "      <th>PED base counts per informative SNP site</th>\n",
       "      <th>COL base counts per informative SNP site</th>\n",
       "      <th>PED base counts per informative SNP site</th>\n",
       "      <th>COL base counts per informative SNP site</th>\n",
       "      <th>PED base counts per informative SNP site</th>\n",
       "      <th>COL base counts per informative SNP site</th>\n",
       "      <th>PED base counts per informative SNP site</th>\n",
       "      <th>COL base counts per informative SNP site</th>\n",
       "      <th>PED base counts per informative SNP site</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>707, 597</td>\n",
       "      <td>0, 0</td>\n",
       "      <td>672, 643</td>\n",
       "      <td>0, 0</td>\n",
       "      <td>668, 661</td>\n",
       "      <td>0, 0</td>\n",
       "      <td>388, 405</td>\n",
       "      <td>0, 0</td>\n",
       "      <td>732, 612</td>\n",
       "      <td>0, 0</td>\n",
       "      <td>...</td>\n",
       "      <td>0, 0</td>\n",
       "      <td>322, 203</td>\n",
       "      <td>0, 0</td>\n",
       "      <td>299, 164</td>\n",
       "      <td>0, 0</td>\n",
       "      <td>310, 192</td>\n",
       "      <td>0, 0</td>\n",
       "      <td>153, 110</td>\n",
       "      <td>4, 0</td>\n",
       "      <td>937, 1178</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>66, 77, 53, 89, 25</td>\n",
       "      <td>0, 0, 0, 0, 0</td>\n",
       "      <td>91, 99, 60, 97, 22</td>\n",
       "      <td>0, 0, 0, 0, 0</td>\n",
       "      <td>93, 89, 59, 98, 45</td>\n",
       "      <td>0, 0, 0, 0, 0</td>\n",
       "      <td>51, 54, 52, 58, 21</td>\n",
       "      <td>0, 0, 0, 0, 0</td>\n",
       "      <td>89, 111, 55, 102, 29</td>\n",
       "      <td>0, 0, 0, 0, 0</td>\n",
       "      <td>...</td>\n",
       "      <td>0, 0, 0, 0, 0</td>\n",
       "      <td>37, 44, 19, 53, 6</td>\n",
       "      <td>0, 0, 0, 0, 0</td>\n",
       "      <td>52, 55, 21, 35, 8</td>\n",
       "      <td>0, 0, 0, 0, 0</td>\n",
       "      <td>44, 61, 33, 56, 15</td>\n",
       "      <td>NA, 0, 0, 0, 0</td>\n",
       "      <td>NA, 17, 14, 11, 3</td>\n",
       "      <td>0, 0, 0, 0, 0</td>\n",
       "      <td>87, 134, 135, 123, 54</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>24, 9, 6</td>\n",
       "      <td>0, 0, 0</td>\n",
       "      <td>30, 7, 5</td>\n",
       "      <td>0, 0, 0</td>\n",
       "      <td>35, 11, 6</td>\n",
       "      <td>0, 0, 0</td>\n",
       "      <td>13, 7, NA</td>\n",
       "      <td>0, 0, NA</td>\n",
       "      <td>17, 8, 6</td>\n",
       "      <td>0, 0, 0</td>\n",
       "      <td>...</td>\n",
       "      <td>0, 0, NA</td>\n",
       "      <td>19, 13, NA</td>\n",
       "      <td>0, 0, NA</td>\n",
       "      <td>21, 10, NA</td>\n",
       "      <td>0, 0, NA</td>\n",
       "      <td>20, 12, NA</td>\n",
       "      <td>0, 0, NA</td>\n",
       "      <td>17, 5, NA</td>\n",
       "      <td>0, 0, 0</td>\n",
       "      <td>131, 125, 30</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>348</td>\n",
       "      <td>0</td>\n",
       "      <td>366</td>\n",
       "      <td>0</td>\n",
       "      <td>329</td>\n",
       "      <td>0</td>\n",
       "      <td>258</td>\n",
       "      <td>0</td>\n",
       "      <td>440</td>\n",
       "      <td>0</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>268</td>\n",
       "      <td>0</td>\n",
       "      <td>291</td>\n",
       "      <td>0</td>\n",
       "      <td>309</td>\n",
       "      <td>0</td>\n",
       "      <td>137</td>\n",
       "      <td>0</td>\n",
       "      <td>1187</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>49, 72, 49, 49, 45, 49, 37, 31, 28, 30, 33, 31...</td>\n",
       "      <td>0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0...</td>\n",
       "      <td>60, 91, 63, 61, 57, 61, 52, 34, 34, 41, 40, 37...</td>\n",
       "      <td>0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0...</td>\n",
       "      <td>53, 101, 76, 73, 58, 63, 52, 36, 33, 33, 30, 3...</td>\n",
       "      <td>0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0...</td>\n",
       "      <td>39, 46, 32, 32, 34, 34, 33, 23, 24, 25, 25, 26...</td>\n",
       "      <td>0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0...</td>\n",
       "      <td>86, 135, 92, 89, 72, 76, 62, 48, 42, 49, 55, 5...</td>\n",
       "      <td>0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0...</td>\n",
       "      <td>...</td>\n",
       "      <td>0, 0, 0, 0, 0, 0, 0, NA, NA, NA, NA, NA, NA, N...</td>\n",
       "      <td>52, 50, 46, 43, 22, 19, 16, NA, NA, NA, NA, NA...</td>\n",
       "      <td>0, 0, 0, 0, 0, 0, NA, NA, NA, NA, NA, 0, 0, 0,...</td>\n",
       "      <td>30, 40, 36, 36, 18, 18, NA, NA, NA, NA, NA, 4,...</td>\n",
       "      <td>0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, NA, 0, 0, ...</td>\n",
       "      <td>32, 41, 42, 35, 24, 17, 13, 4, 4, 4, 6, 3, NA,...</td>\n",
       "      <td>0, 0, 0, 0, NA, NA, 0, NA, NA, NA, 0, 0, 0, 0,...</td>\n",
       "      <td>15, 21, 24, 22, NA, NA, 10, NA, NA, NA, 8, 7, ...</td>\n",
       "      <td>0, 0, 0, 0, 0, 0, 0, NA, NA, NA, 0, 0, 0, 0, 0...</td>\n",
       "      <td>86, 103, 93, 84, 52, 52, 18, NA, NA, NA, 5, 3,...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12062</th>\n",
       "      <td>159, 249, 219, 395, 1154, 1382, 1386, 574, 919...</td>\n",
       "      <td>0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0...</td>\n",
       "      <td>197, 287, 240, 364, 1213, 1415, 1438, 590, 867...</td>\n",
       "      <td>0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0...</td>\n",
       "      <td>306, 449, 361, 387, 1363, 1716, 1779, 627, 981...</td>\n",
       "      <td>0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0...</td>\n",
       "      <td>63, 96, 79, 156, 493, 578, 617, 239, 364, 363,...</td>\n",
       "      <td>0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0...</td>\n",
       "      <td>138, 205, 157, 226, 772, 959, 985, 363, 618, 6...</td>\n",
       "      <td>0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0...</td>\n",
       "      <td>...</td>\n",
       "      <td>NA, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, ...</td>\n",
       "      <td>NA, 43, 31, 134, 377, 295, 294, 93, 28, 79, 12...</td>\n",
       "      <td>0, 0, NA, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, ...</td>\n",
       "      <td>33, 43, NA, 87, 271, 242, 243, 89, 30, 97, 144...</td>\n",
       "      <td>NA, 0, NA, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0,...</td>\n",
       "      <td>NA, 61, NA, 122, 334, 266, 268, 88, 39, 124, 1...</td>\n",
       "      <td>NA, 0, NA, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0,...</td>\n",
       "      <td>NA, 44, NA, 71, 232, 182, 196, 56, 24, 64, 99,...</td>\n",
       "      <td>0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0...</td>\n",
       "      <td>52, 43, 49, 237, 530, 579, 494, 649, 197, 377,...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12063</th>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>6</td>\n",
       "      <td>0</td>\n",
       "      <td>5</td>\n",
       "      <td>0</td>\n",
       "      <td>3</td>\n",
       "      <td>0</td>\n",
       "      <td>6</td>\n",
       "      <td>0</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>3</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12064</th>\n",
       "      <td>4, NA</td>\n",
       "      <td>0, NA</td>\n",
       "      <td>6, NA</td>\n",
       "      <td>0, NA</td>\n",
       "      <td>3, NA</td>\n",
       "      <td>0, NA</td>\n",
       "      <td>141, 123</td>\n",
       "      <td>0, 0</td>\n",
       "      <td>244, 234</td>\n",
       "      <td>0, 0</td>\n",
       "      <td>...</td>\n",
       "      <td>0, 0</td>\n",
       "      <td>567, 373</td>\n",
       "      <td>0, 0</td>\n",
       "      <td>572, 437</td>\n",
       "      <td>0, 0</td>\n",
       "      <td>443, 304</td>\n",
       "      <td>0, 0</td>\n",
       "      <td>1705, 1255</td>\n",
       "      <td>0, 3</td>\n",
       "      <td>12, 8</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12065</th>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>28</td>\n",
       "      <td>0</td>\n",
       "      <td>30</td>\n",
       "      <td>0</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>275</td>\n",
       "      <td>0</td>\n",
       "      <td>196</td>\n",
       "      <td>0</td>\n",
       "      <td>301</td>\n",
       "      <td>0</td>\n",
       "      <td>1048</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12066</th>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>3</td>\n",
       "      <td>0</td>\n",
       "      <td>27</td>\n",
       "      <td>0</td>\n",
       "      <td>32</td>\n",
       "      <td>0</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>58</td>\n",
       "      <td>0</td>\n",
       "      <td>57</td>\n",
       "      <td>0</td>\n",
       "      <td>92</td>\n",
       "      <td>0</td>\n",
       "      <td>280</td>\n",
       "      <td>0</td>\n",
       "      <td>16</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>12067 rows × 58 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                                                                                                Cr*Ps r1  \\\n",
       "      COL base counts per informative SNP site (NA in cases of no read confirming either of two alleles)   \n",
       "0                                               707, 597                                                   \n",
       "1                                     66, 77, 53, 89, 25                                                   \n",
       "2                                               24, 9, 6                                                   \n",
       "3                                                    348                                                   \n",
       "4      49, 72, 49, 49, 45, 49, 37, 31, 28, 30, 33, 31...                                                   \n",
       "...                                                  ...                                                   \n",
       "12062  159, 249, 219, 395, 1154, 1382, 1386, 574, 919...                                                   \n",
       "12063                                                NaN                                                   \n",
       "12064                                              4, NA                                                   \n",
       "12065                                                NaN                                                   \n",
       "12066                                                NaN                                                   \n",
       "\n",
       "                                                          \\\n",
       "                PED base counts per informative SNP site   \n",
       "0                                                   0, 0   \n",
       "1                                          0, 0, 0, 0, 0   \n",
       "2                                                0, 0, 0   \n",
       "3                                                      0   \n",
       "4      0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0...   \n",
       "...                                                  ...   \n",
       "12062  0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0...   \n",
       "12063                                                NaN   \n",
       "12064                                              0, NA   \n",
       "12065                                                NaN   \n",
       "12066                                                NaN   \n",
       "\n",
       "                                                Cr*Ps r2  \\\n",
       "                COL base counts per informative SNP site   \n",
       "0                                               672, 643   \n",
       "1                                     91, 99, 60, 97, 22   \n",
       "2                                               30, 7, 5   \n",
       "3                                                    366   \n",
       "4      60, 91, 63, 61, 57, 61, 52, 34, 34, 41, 40, 37...   \n",
       "...                                                  ...   \n",
       "12062  197, 287, 240, 364, 1213, 1415, 1438, 590, 867...   \n",
       "12063                                                  6   \n",
       "12064                                              6, NA   \n",
       "12065                                                NaN   \n",
       "12066                                                NaN   \n",
       "\n",
       "                                                          \\\n",
       "                PED base counts per informative SNP site   \n",
       "0                                                   0, 0   \n",
       "1                                          0, 0, 0, 0, 0   \n",
       "2                                                0, 0, 0   \n",
       "3                                                      0   \n",
       "4      0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0...   \n",
       "...                                                  ...   \n",
       "12062  0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0...   \n",
       "12063                                                  0   \n",
       "12064                                              0, NA   \n",
       "12065                                                NaN   \n",
       "12066                                                NaN   \n",
       "\n",
       "                                                Cr*Ps r3  \\\n",
       "                COL base counts per informative SNP site   \n",
       "0                                               668, 661   \n",
       "1                                     93, 89, 59, 98, 45   \n",
       "2                                              35, 11, 6   \n",
       "3                                                    329   \n",
       "4      53, 101, 76, 73, 58, 63, 52, 36, 33, 33, 30, 3...   \n",
       "...                                                  ...   \n",
       "12062  306, 449, 361, 387, 1363, 1716, 1779, 627, 981...   \n",
       "12063                                                  5   \n",
       "12064                                              3, NA   \n",
       "12065                                                NaN   \n",
       "12066                                                  3   \n",
       "\n",
       "                                                          \\\n",
       "                PED base counts per informative SNP site   \n",
       "0                                                   0, 0   \n",
       "1                                          0, 0, 0, 0, 0   \n",
       "2                                                0, 0, 0   \n",
       "3                                                      0   \n",
       "4      0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0...   \n",
       "...                                                  ...   \n",
       "12062  0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0...   \n",
       "12063                                                  0   \n",
       "12064                                              0, NA   \n",
       "12065                                                NaN   \n",
       "12066                                                  0   \n",
       "\n",
       "                                                PrCs* r1  \\\n",
       "                COL base counts per informative SNP site   \n",
       "0                                               388, 405   \n",
       "1                                     51, 54, 52, 58, 21   \n",
       "2                                              13, 7, NA   \n",
       "3                                                    258   \n",
       "4      39, 46, 32, 32, 34, 34, 33, 23, 24, 25, 25, 26...   \n",
       "...                                                  ...   \n",
       "12062  63, 96, 79, 156, 493, 578, 617, 239, 364, 363,...   \n",
       "12063                                                  3   \n",
       "12064                                           141, 123   \n",
       "12065                                                 28   \n",
       "12066                                                 27   \n",
       "\n",
       "                                                          \\\n",
       "                PED base counts per informative SNP site   \n",
       "0                                                   0, 0   \n",
       "1                                          0, 0, 0, 0, 0   \n",
       "2                                               0, 0, NA   \n",
       "3                                                      0   \n",
       "4      0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0...   \n",
       "...                                                  ...   \n",
       "12062  0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0...   \n",
       "12063                                                  0   \n",
       "12064                                               0, 0   \n",
       "12065                                                  0   \n",
       "12066                                                  0   \n",
       "\n",
       "                                                PrCs* r2  \\\n",
       "                COL base counts per informative SNP site   \n",
       "0                                               732, 612   \n",
       "1                                   89, 111, 55, 102, 29   \n",
       "2                                               17, 8, 6   \n",
       "3                                                    440   \n",
       "4      86, 135, 92, 89, 72, 76, 62, 48, 42, 49, 55, 5...   \n",
       "...                                                  ...   \n",
       "12062  138, 205, 157, 226, 772, 959, 985, 363, 618, 6...   \n",
       "12063                                                  6   \n",
       "12064                                           244, 234   \n",
       "12065                                                 30   \n",
       "12066                                                 32   \n",
       "\n",
       "                                                          ...  \\\n",
       "                PED base counts per informative SNP site  ...   \n",
       "0                                                   0, 0  ...   \n",
       "1                                          0, 0, 0, 0, 0  ...   \n",
       "2                                                0, 0, 0  ...   \n",
       "3                                                      0  ...   \n",
       "4      0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0...  ...   \n",
       "...                                                  ...  ...   \n",
       "12062  0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0...  ...   \n",
       "12063                                                  0  ...   \n",
       "12064                                               0, 0  ...   \n",
       "12065                                                  0  ...   \n",
       "12066                                                  0  ...   \n",
       "\n",
       "                                             CrPs* -N r3  \\\n",
       "                COL base counts per informative SNP site   \n",
       "0                                                   0, 0   \n",
       "1                                          0, 0, 0, 0, 0   \n",
       "2                                               0, 0, NA   \n",
       "3                                                      0   \n",
       "4      0, 0, 0, 0, 0, 0, 0, NA, NA, NA, NA, NA, NA, N...   \n",
       "...                                                  ...   \n",
       "12062  NA, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, ...   \n",
       "12063                                                  0   \n",
       "12064                                               0, 0   \n",
       "12065                                                  0   \n",
       "12066                                                  0   \n",
       "\n",
       "                                                          \\\n",
       "                PED base counts per informative SNP site   \n",
       "0                                               322, 203   \n",
       "1                                      37, 44, 19, 53, 6   \n",
       "2                                             19, 13, NA   \n",
       "3                                                    268   \n",
       "4      52, 50, 46, 43, 22, 19, 16, NA, NA, NA, NA, NA...   \n",
       "...                                                  ...   \n",
       "12062  NA, 43, 31, 134, 377, 295, 294, 93, 28, 79, 12...   \n",
       "12063                                                  3   \n",
       "12064                                           567, 373   \n",
       "12065                                                275   \n",
       "12066                                                 58   \n",
       "\n",
       "                                             CrPs* -P r1  \\\n",
       "                COL base counts per informative SNP site   \n",
       "0                                                   0, 0   \n",
       "1                                          0, 0, 0, 0, 0   \n",
       "2                                               0, 0, NA   \n",
       "3                                                      0   \n",
       "4      0, 0, 0, 0, 0, 0, NA, NA, NA, NA, NA, 0, 0, 0,...   \n",
       "...                                                  ...   \n",
       "12062  0, 0, NA, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, ...   \n",
       "12063                                                NaN   \n",
       "12064                                               0, 0   \n",
       "12065                                                  0   \n",
       "12066                                                  0   \n",
       "\n",
       "                                                          \\\n",
       "                PED base counts per informative SNP site   \n",
       "0                                               299, 164   \n",
       "1                                      52, 55, 21, 35, 8   \n",
       "2                                             21, 10, NA   \n",
       "3                                                    291   \n",
       "4      30, 40, 36, 36, 18, 18, NA, NA, NA, NA, NA, 4,...   \n",
       "...                                                  ...   \n",
       "12062  33, 43, NA, 87, 271, 242, 243, 89, 30, 97, 144...   \n",
       "12063                                                NaN   \n",
       "12064                                           572, 437   \n",
       "12065                                                196   \n",
       "12066                                                 57   \n",
       "\n",
       "                                             CrPs* -P r2  \\\n",
       "                COL base counts per informative SNP site   \n",
       "0                                                   0, 0   \n",
       "1                                          0, 0, 0, 0, 0   \n",
       "2                                               0, 0, NA   \n",
       "3                                                      0   \n",
       "4      0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, NA, 0, 0, ...   \n",
       "...                                                  ...   \n",
       "12062  NA, 0, NA, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0,...   \n",
       "12063                                                NaN   \n",
       "12064                                               0, 0   \n",
       "12065                                                  0   \n",
       "12066                                                  0   \n",
       "\n",
       "                                                          \\\n",
       "                PED base counts per informative SNP site   \n",
       "0                                               310, 192   \n",
       "1                                     44, 61, 33, 56, 15   \n",
       "2                                             20, 12, NA   \n",
       "3                                                    309   \n",
       "4      32, 41, 42, 35, 24, 17, 13, 4, 4, 4, 6, 3, NA,...   \n",
       "...                                                  ...   \n",
       "12062  NA, 61, NA, 122, 334, 266, 268, 88, 39, 124, 1...   \n",
       "12063                                                NaN   \n",
       "12064                                           443, 304   \n",
       "12065                                                301   \n",
       "12066                                                 92   \n",
       "\n",
       "                                             CrPs* -P r3  \\\n",
       "                COL base counts per informative SNP site   \n",
       "0                                                   0, 0   \n",
       "1                                         NA, 0, 0, 0, 0   \n",
       "2                                               0, 0, NA   \n",
       "3                                                      0   \n",
       "4      0, 0, 0, 0, NA, NA, 0, NA, NA, NA, 0, 0, 0, 0,...   \n",
       "...                                                  ...   \n",
       "12062  NA, 0, NA, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0,...   \n",
       "12063                                                NaN   \n",
       "12064                                               0, 0   \n",
       "12065                                                  0   \n",
       "12066                                                  0   \n",
       "\n",
       "                                                          \\\n",
       "                PED base counts per informative SNP site   \n",
       "0                                               153, 110   \n",
       "1                                      NA, 17, 14, 11, 3   \n",
       "2                                              17, 5, NA   \n",
       "3                                                    137   \n",
       "4      15, 21, 24, 22, NA, NA, 10, NA, NA, NA, 8, 7, ...   \n",
       "...                                                  ...   \n",
       "12062  NA, 44, NA, 71, 232, 182, 196, 56, 24, 64, 99,...   \n",
       "12063                                                NaN   \n",
       "12064                                         1705, 1255   \n",
       "12065                                               1048   \n",
       "12066                                                280   \n",
       "\n",
       "                                           Pr*Cs FN_root  \\\n",
       "                COL base counts per informative SNP site   \n",
       "0                                                   4, 0   \n",
       "1                                          0, 0, 0, 0, 0   \n",
       "2                                                0, 0, 0   \n",
       "3                                                      0   \n",
       "4      0, 0, 0, 0, 0, 0, 0, NA, NA, NA, 0, 0, 0, 0, 0...   \n",
       "...                                                  ...   \n",
       "12062  0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0...   \n",
       "12063                                                NaN   \n",
       "12064                                               0, 3   \n",
       "12065                                                NaN   \n",
       "12066                                                  0   \n",
       "\n",
       "                                                          \n",
       "                PED base counts per informative SNP site  \n",
       "0                                              937, 1178  \n",
       "1                                  87, 134, 135, 123, 54  \n",
       "2                                           131, 125, 30  \n",
       "3                                                   1187  \n",
       "4      86, 103, 93, 84, 52, 52, 18, NA, NA, NA, 5, 3,...  \n",
       "...                                                  ...  \n",
       "12062  52, 43, 49, 237, 530, 579, 494, 649, 197, 377,...  \n",
       "12063                                                NaN  \n",
       "12064                                              12, 8  \n",
       "12065                                                NaN  \n",
       "12066                                                 16  \n",
       "\n",
       "[12067 rows x 58 columns]"
      ]
     },
     "execution_count": 6,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "## Convert df into a multi-level dataframe\n",
    "import pandas as pd\n",
    "df = pd.read_excel(\"thieme_original_data.xlsx\", sheet_name=\"SNP in all transcripts\", header=[0,1])\n",
    "cols = df.columns.to_list()\n",
    "exp_cols = [col for col in cols if \"*\" in col[0]]\n",
    "## For each set up, we only want to keep COL base and PED base\n",
    "## loop through the exp_cols list and keep those that contain \"PED base counts\" or \"COL base counts\"\n",
    "\n",
    "exp_cols_cp = [col for col in cols if \"COL base \" in col[1] or \"PED base \" in col[1]]\n",
    "df_exps = df[exp_cols_cp]\n",
    "\n",
    "df_exps\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "['Cr*Ps r1',\n",
       " 'PrCs* FN_flower',\n",
       " 'PrCs* r1',\n",
       " 'Pr*Cs r1',\n",
       " 'CrPs* -N r1',\n",
       " 'CrPs* -P r1',\n",
       " 'CrPs* r2',\n",
       " 'PrCs* FN_stemUpper',\n",
       " 'CrPs* -N r3',\n",
       " 'CrPs* -P r2',\n",
       " 'Cr*Ps -N r3',\n",
       " 'Cr*Ps -N r1',\n",
       " 'Cr*Ps r3',\n",
       " 'PrCs* r2',\n",
       " 'CrPs* r3',\n",
       " 'CrPs* r1',\n",
       " 'Cr*Ps r2',\n",
       " 'Cr*Ps -N r2',\n",
       " 'PrCs* r3',\n",
       " 'Pr*Cs FN_root',\n",
       " 'Pr*Cs r2',\n",
       " 'PrCs* FN_stemLower',\n",
       " 'Pr*Cs r3',\n",
       " 'CrPs* -N r2',\n",
       " 'Cr*Ps -P r1',\n",
       " 'Cr*Ps -P r3',\n",
       " 'CrPs* -P r3',\n",
       " 'PrCs* FN_rosette',\n",
       " 'Cr*Ps -P r2']"
      ]
     },
     "execution_count": 7,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "## Get just the experiments\n",
    "exp_list = [exp[0] for exp in exp_cols]\n",
    "exp_list = list(set(exp_list))\n",
    "exp_list"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Transcript</th>\n",
       "      <th>Informative positions</th>\n",
       "      <th>COL base</th>\n",
       "      <th>PED base</th>\n",
       "      <th>rep</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>AT1G50920.1</td>\n",
       "      <td>220, 1351</td>\n",
       "      <td>707, 597</td>\n",
       "      <td>0, 0</td>\n",
       "      <td>Cr*Ps r1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>AT1G73440.1</td>\n",
       "      <td>122, 158, 345, 421, 813</td>\n",
       "      <td>66, 77, 53, 89, 25</td>\n",
       "      <td>0, 0, 0, 0, 0</td>\n",
       "      <td>Cr*Ps r1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>AT1G51380.1</td>\n",
       "      <td>209, 987, 1376</td>\n",
       "      <td>24, 9, 6</td>\n",
       "      <td>0, 0, 0</td>\n",
       "      <td>Cr*Ps r1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>AT1G10950.1</td>\n",
       "      <td>221</td>\n",
       "      <td>348</td>\n",
       "      <td>0</td>\n",
       "      <td>Cr*Ps r1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>AT1G31870.1</td>\n",
       "      <td>100, 246, 273, 289, 363, 368, 395, 457, 469, 4...</td>\n",
       "      <td>49, 72, 49, 49, 45, 49, 37, 31, 28, 30, 33, 31...</td>\n",
       "      <td>0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0...</td>\n",
       "      <td>Cr*Ps r1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12062</th>\n",
       "      <td>AT5G25560.4</td>\n",
       "      <td>108, 189, 231, 417, 469, 593, 614, 715, 971, 9...</td>\n",
       "      <td>159, 249, 219, 395, 1154, 1382, 1386, 574, 919...</td>\n",
       "      <td>0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0...</td>\n",
       "      <td>Cr*Ps r1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12063</th>\n",
       "      <td>AT5G45790.2</td>\n",
       "      <td>1252</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Cr*Ps r1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12064</th>\n",
       "      <td>ATCG00490.1</td>\n",
       "      <td>864, 1116</td>\n",
       "      <td>4, NA</td>\n",
       "      <td>0, NA</td>\n",
       "      <td>Cr*Ps r1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12065</th>\n",
       "      <td>ATCG00680.1</td>\n",
       "      <td>222</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Cr*Ps r1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12066</th>\n",
       "      <td>ATCG00540.1</td>\n",
       "      <td>629</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Cr*Ps r1</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>12067 rows × 5 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "        Transcript                              Informative positions  \\\n",
       "0      AT1G50920.1                                          220, 1351   \n",
       "1      AT1G73440.1                            122, 158, 345, 421, 813   \n",
       "2      AT1G51380.1                                     209, 987, 1376   \n",
       "3      AT1G10950.1                                                221   \n",
       "4      AT1G31870.1  100, 246, 273, 289, 363, 368, 395, 457, 469, 4...   \n",
       "...            ...                                                ...   \n",
       "12062  AT5G25560.4  108, 189, 231, 417, 469, 593, 614, 715, 971, 9...   \n",
       "12063  AT5G45790.2                                               1252   \n",
       "12064  ATCG00490.1                                          864, 1116   \n",
       "12065  ATCG00680.1                                                222   \n",
       "12066  ATCG00540.1                                                629   \n",
       "\n",
       "                                                COL base  \\\n",
       "0                                               707, 597   \n",
       "1                                     66, 77, 53, 89, 25   \n",
       "2                                               24, 9, 6   \n",
       "3                                                    348   \n",
       "4      49, 72, 49, 49, 45, 49, 37, 31, 28, 30, 33, 31...   \n",
       "...                                                  ...   \n",
       "12062  159, 249, 219, 395, 1154, 1382, 1386, 574, 919...   \n",
       "12063                                                NaN   \n",
       "12064                                              4, NA   \n",
       "12065                                                NaN   \n",
       "12066                                                NaN   \n",
       "\n",
       "                                                PED base       rep  \n",
       "0                                                   0, 0  Cr*Ps r1  \n",
       "1                                          0, 0, 0, 0, 0  Cr*Ps r1  \n",
       "2                                                0, 0, 0  Cr*Ps r1  \n",
       "3                                                      0  Cr*Ps r1  \n",
       "4      0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0...  Cr*Ps r1  \n",
       "...                                                  ...       ...  \n",
       "12062  0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0...  Cr*Ps r1  \n",
       "12063                                                NaN  Cr*Ps r1  \n",
       "12064                                              0, NA  Cr*Ps r1  \n",
       "12065                                                NaN  Cr*Ps r1  \n",
       "12066                                                NaN  Cr*Ps r1  \n",
       "\n",
       "[12067 rows x 5 columns]"
      ]
     },
     "execution_count": 8,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "## Add in the transcript information\n",
    "df_info = df[[\"Transcript\",\"Informative positions\"]]\n",
    "## Drop level 1\n",
    "df_info.columns = df_info.columns.droplevel(1)\n",
    "\n",
    "df_list = []\n",
    "for exp in exp_list:\n",
    "    df_subset = df_exps[exp].copy()\n",
    "    df_subset[\"rep\"] = exp\n",
    "    df_subset = pd.concat([df_info, df_subset], axis=1)\n",
    "    cols = df_subset.columns.to_list()\n",
    "    col_col = [col for col in cols if \"COL base\" in col]\n",
    "    col_ped = [col for col in cols if \"PED base\" in col]\n",
    "    df_subset.rename(columns = {col_col[0]:\"COL base\", col_ped[0]:\"PED base\"}, inplace=True)\n",
    "    df_list.append(df_subset)\n",
    "\n",
    "df_list[0]\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {},
   "outputs": [],
   "source": [
    "## Convert the 3 middle columns to lists and explode the dataframe\n",
    "def convert_to_list(row):\n",
    "    snp_list = []\n",
    "    ## Separate into ints\n",
    "    ## Check if there are multiple values\n",
    "    if \",\" in str(row):\n",
    "        snp_list = row.split(\",\")\n",
    "    else:\n",
    "        try:\n",
    "            snp_list = [int(row)]\n",
    "        except:\n",
    "            return [\"NA\"]\n",
    "\n",
    "    return snp_list\n",
    "\n",
    "for df in df_list:\n",
    "    df[\"COL base\"] = df[\"COL base\"].apply(lambda x: convert_to_list(x))\n",
    "    df[\"PED base\"] = df[\"PED base\"].apply(lambda x: convert_to_list(x))\n",
    "    df[\"Informative positions\"] = df[\"Informative positions\"].apply(lambda x: convert_to_list(x))\n",
    "    df.explode([\"COL base\",\"PED base\",\"Informative positions\"])\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "29"
      ]
     },
     "execution_count": 10,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "len(df_list)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {},
   "outputs": [],
   "source": [
    "df_list_new = []\n",
    "for df in df_list:\n",
    "    df_list_new.append(df.explode([\"Informative positions\",\"COL base\",\"PED base\"]))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "29"
      ]
     },
     "execution_count": 12,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "len(df_list_new)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "metadata": {},
   "outputs": [],
   "source": [
    "## Add in the SNP information\n",
    "df_snp = pd.read_csv(\"new_thieme_og_snps.csv\", index_col=None)\n",
    "df_list_snps = []\n",
    "\n",
    "for df in df_list_new:\n",
    "    df[\"Informative positions\"] = pd.to_numeric(df[\"Informative positions\"], errors='coerce')\n",
    "    new_df = pd.merge(df, df_snp,  left_on=[\"Transcript\",\"Informative positions\"], right_on = [\"Transcript\",\"snp_list\"])\n",
    "    new_df[\"COL base\"] = pd.to_numeric(new_df[\"COL base\"], errors='coerce')\n",
    "    new_df[\"PED base\"] = pd.to_numeric(new_df[\"PED base\"], errors='coerce')\n",
    "    ## Let's update the column names in a slightly cheating fashion\n",
    "    if new_df[\"COL base\"].sum() > new_df[\"PED base\"].sum():\n",
    "        new_df.rename(columns = {\"COL base\":\"N\",\"PED base\":\"n\"}, inplace=True)\n",
    "        ## Get if its root or shoot\n",
    "        new_df[\"type\"] = \"Col\"\n",
    "        if \"Cs\" in new_df[\"rep\"][0]:\n",
    "            new_df[\"tissue\"] = \"shoot\"\n",
    "        if \"Cr\" in new_df[\"rep\"][0]:\n",
    "            new_df[\"tissue\"] = \"root\"\n",
    "    else:\n",
    "        new_df.rename(columns = {\"PED base\":\"N\",\"COL base\":\"n\"}, inplace=True)\n",
    "        new_df[\"type\"] = \"Ped\"\n",
    "        if \"Ps\" in new_df[\"rep\"][0]:\n",
    "            new_df[\"tissue\"] = \"shoot\"\n",
    "        if \"Pr\" in new_df[\"rep\"][0]:\n",
    "            new_df[\"tissue\"] = \"root\"\n",
    "    df_list_snps.append(new_df)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Transcript</th>\n",
       "      <th>Informative positions</th>\n",
       "      <th>N</th>\n",
       "      <th>n</th>\n",
       "      <th>rep</th>\n",
       "      <th>snp_list</th>\n",
       "      <th>SNP</th>\n",
       "      <th>type</th>\n",
       "      <th>tissue</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>AT1G50920.1</td>\n",
       "      <td>220</td>\n",
       "      <td>707.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>Cr*Ps r1</td>\n",
       "      <td>220</td>\n",
       "      <td>Chr1_18870656</td>\n",
       "      <td>Col</td>\n",
       "      <td>root</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>AT1G50920.1</td>\n",
       "      <td>1351</td>\n",
       "      <td>597.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>Cr*Ps r1</td>\n",
       "      <td>1351</td>\n",
       "      <td>Chr1_18871787</td>\n",
       "      <td>Col</td>\n",
       "      <td>root</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>AT1G51380.1</td>\n",
       "      <td>209</td>\n",
       "      <td>24.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>Cr*Ps r1</td>\n",
       "      <td>209</td>\n",
       "      <td>Chr1_19048090</td>\n",
       "      <td>Col</td>\n",
       "      <td>root</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>AT1G51380.1</td>\n",
       "      <td>987</td>\n",
       "      <td>9.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>Cr*Ps r1</td>\n",
       "      <td>987</td>\n",
       "      <td>Chr1_19049605</td>\n",
       "      <td>Col</td>\n",
       "      <td>root</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>AT1G51380.1</td>\n",
       "      <td>1376</td>\n",
       "      <td>6.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>Cr*Ps r1</td>\n",
       "      <td>1376</td>\n",
       "      <td>Chr1_19050086</td>\n",
       "      <td>Col</td>\n",
       "      <td>root</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>54648</th>\n",
       "      <td>AT5G25560.4</td>\n",
       "      <td>2225</td>\n",
       "      <td>61.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>Cr*Ps r1</td>\n",
       "      <td>2225</td>\n",
       "      <td>Chr5_8901116</td>\n",
       "      <td>Col</td>\n",
       "      <td>root</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>54649</th>\n",
       "      <td>AT5G25560.4</td>\n",
       "      <td>2240</td>\n",
       "      <td>62.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>Cr*Ps r1</td>\n",
       "      <td>2240</td>\n",
       "      <td>Chr5_8901131</td>\n",
       "      <td>Col</td>\n",
       "      <td>root</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>54650</th>\n",
       "      <td>AT5G25560.4</td>\n",
       "      <td>2249</td>\n",
       "      <td>53.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>Cr*Ps r1</td>\n",
       "      <td>2249</td>\n",
       "      <td>Chr5_8901140</td>\n",
       "      <td>Col</td>\n",
       "      <td>root</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>54651</th>\n",
       "      <td>AT5G25560.4</td>\n",
       "      <td>2252</td>\n",
       "      <td>49.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>Cr*Ps r1</td>\n",
       "      <td>2252</td>\n",
       "      <td>Chr5_8901143</td>\n",
       "      <td>Col</td>\n",
       "      <td>root</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>54652</th>\n",
       "      <td>AT5G25560.4</td>\n",
       "      <td>2336</td>\n",
       "      <td>44.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>Cr*Ps r1</td>\n",
       "      <td>2336</td>\n",
       "      <td>Chr5_8901310</td>\n",
       "      <td>Col</td>\n",
       "      <td>root</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>54653 rows × 9 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "        Transcript  Informative positions      N    n       rep  snp_list  \\\n",
       "0      AT1G50920.1                    220  707.0  0.0  Cr*Ps r1       220   \n",
       "1      AT1G50920.1                   1351  597.0  0.0  Cr*Ps r1      1351   \n",
       "2      AT1G51380.1                    209   24.0  0.0  Cr*Ps r1       209   \n",
       "3      AT1G51380.1                    987    9.0  0.0  Cr*Ps r1       987   \n",
       "4      AT1G51380.1                   1376    6.0  0.0  Cr*Ps r1      1376   \n",
       "...            ...                    ...    ...  ...       ...       ...   \n",
       "54648  AT5G25560.4                   2225   61.0  0.0  Cr*Ps r1      2225   \n",
       "54649  AT5G25560.4                   2240   62.0  0.0  Cr*Ps r1      2240   \n",
       "54650  AT5G25560.4                   2249   53.0  0.0  Cr*Ps r1      2249   \n",
       "54651  AT5G25560.4                   2252   49.0  0.0  Cr*Ps r1      2252   \n",
       "54652  AT5G25560.4                   2336   44.0  0.0  Cr*Ps r1      2336   \n",
       "\n",
       "                 SNP type tissue  \n",
       "0      Chr1_18870656  Col   root  \n",
       "1      Chr1_18871787  Col   root  \n",
       "2      Chr1_19048090  Col   root  \n",
       "3      Chr1_19049605  Col   root  \n",
       "4      Chr1_19050086  Col   root  \n",
       "...              ...  ...    ...  \n",
       "54648   Chr5_8901116  Col   root  \n",
       "54649   Chr5_8901131  Col   root  \n",
       "54650   Chr5_8901140  Col   root  \n",
       "54651   Chr5_8901143  Col   root  \n",
       "54652   Chr5_8901310  Col   root  \n",
       "\n",
       "[54653 rows x 9 columns]"
      ]
     },
     "execution_count": 14,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df_list_snps[0]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "metadata": {},
   "outputs": [],
   "source": [
    "## Time to add in the homograft data! Huzzah!\n",
    "\n",
    "ccroot = pd.read_csv(\"raw_data/thieme/homfiles/C-C-Root-FN.txt\",delimiter = \"\\t\", low_memory = False)\n",
    "pproot = pd.read_csv(\"raw_data/thieme/homfiles/P-P-Root-FN.txt\",delimiter = \"\\t\", low_memory = False)\n",
    "ccshoot = pd.read_csv(\"raw_data/thieme/homfiles/C-C-Shoot-FN.txt\",delimiter = \"\\t\", low_memory = False)\n",
    "ppshoot = pd.read_csv(\"raw_data/thieme/homfiles/P-P-Shoot-FN.txt\",delimiter = \"\\t\", low_memory = False)\n",
    "\n",
    "ccroot['depth'] = pd.to_numeric(ccroot['depth'], errors='coerce')\n",
    "ccroot['lerDepth'] = pd.to_numeric(ccroot['lerDepth'], errors='coerce')\n",
    "ccroot['rawDepth'] = pd.to_numeric(ccroot['rawDepth'], errors='coerce')\n",
    "\n",
    "ccroot = ccroot.dropna(subset=['depth', 'lerDepth', 'rawDepth'])\n",
    "ccroot = ccroot[ccroot[\"depth\"]>0]\n",
    "\n",
    "## Repeat all the above code for pproot\n",
    "pproot['depth'] = pd.to_numeric(pproot['depth'], errors='coerce')\n",
    "pproot['colDepth'] = pd.to_numeric(pproot['colDepth'], errors='coerce')\n",
    "pproot['rawDepth'] = pd.to_numeric(pproot['rawDepth'], errors='coerce')\n",
    "\n",
    "pproot = pproot.dropna(subset=['depth', 'colDepth', 'rawDepth'])\n",
    "pproot = pproot[pproot[\"depth\"]>0]\n",
    "\n",
    "## Repeat above for ccshoot\n",
    "\n",
    "ccshoot['depth'] = pd.to_numeric(ccshoot['depth'], errors='coerce')\n",
    "ccshoot['lerDepth'] = pd.to_numeric(ccshoot['lerDepth'], errors='coerce')\n",
    "ccshoot['rawDepth'] = pd.to_numeric(ccshoot['rawDepth'], errors='coerce')\n",
    "\n",
    "ccshoot = ccshoot.dropna(subset=['depth', 'lerDepth', 'rawDepth'])\n",
    "ccshoot = ccshoot[ccshoot[\"depth\"]>0]\n",
    "\n",
    "## Repeat above for ppshoot\n",
    "\n",
    "ppshoot['depth'] = pd.to_numeric(ppshoot['depth'], errors='coerce')\n",
    "ppshoot['colDepth'] = pd.to_numeric(ppshoot['colDepth'], errors='coerce')\n",
    "ppshoot['rawDepth'] = pd.to_numeric(ppshoot['rawDepth'], errors='coerce')\n",
    "\n",
    "ppshoot = ppshoot.dropna(subset=['depth', 'colDepth', 'rawDepth'])\n",
    "ppshoot = ppshoot[ppshoot[\"depth\"]>0]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "metadata": {},
   "outputs": [],
   "source": [
    "## Need to sum across the replicates before adding the homograft data\n",
    "df_bayes_exp = []\n",
    "for df in df_list_snps:\n",
    "    if \"FN\" not in df[\"rep\"][0]:\n",
    "        df[\"exp\"] = df[\"rep\"].apply(lambda x: x.split(\" \")[0])\n",
    "    else:\n",
    "        df[\"exp\"] = df[\"rep\"]\n",
    "    df_bayes_exp.append(df)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Transcript</th>\n",
       "      <th>Informative positions</th>\n",
       "      <th>N</th>\n",
       "      <th>n</th>\n",
       "      <th>rep</th>\n",
       "      <th>snp_list</th>\n",
       "      <th>SNP</th>\n",
       "      <th>type</th>\n",
       "      <th>tissue</th>\n",
       "      <th>exp</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>AT1G50920.1</td>\n",
       "      <td>220</td>\n",
       "      <td>707.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>Cr*Ps r1</td>\n",
       "      <td>220</td>\n",
       "      <td>Chr1_18870656</td>\n",
       "      <td>Col</td>\n",
       "      <td>root</td>\n",
       "      <td>Cr*Ps</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>AT1G50920.1</td>\n",
       "      <td>1351</td>\n",
       "      <td>597.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>Cr*Ps r1</td>\n",
       "      <td>1351</td>\n",
       "      <td>Chr1_18871787</td>\n",
       "      <td>Col</td>\n",
       "      <td>root</td>\n",
       "      <td>Cr*Ps</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>AT1G51380.1</td>\n",
       "      <td>209</td>\n",
       "      <td>24.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>Cr*Ps r1</td>\n",
       "      <td>209</td>\n",
       "      <td>Chr1_19048090</td>\n",
       "      <td>Col</td>\n",
       "      <td>root</td>\n",
       "      <td>Cr*Ps</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>AT1G51380.1</td>\n",
       "      <td>987</td>\n",
       "      <td>9.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>Cr*Ps r1</td>\n",
       "      <td>987</td>\n",
       "      <td>Chr1_19049605</td>\n",
       "      <td>Col</td>\n",
       "      <td>root</td>\n",
       "      <td>Cr*Ps</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>AT1G51380.1</td>\n",
       "      <td>1376</td>\n",
       "      <td>6.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>Cr*Ps r1</td>\n",
       "      <td>1376</td>\n",
       "      <td>Chr1_19050086</td>\n",
       "      <td>Col</td>\n",
       "      <td>root</td>\n",
       "      <td>Cr*Ps</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>54648</th>\n",
       "      <td>AT5G25560.4</td>\n",
       "      <td>2225</td>\n",
       "      <td>61.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>Cr*Ps r1</td>\n",
       "      <td>2225</td>\n",
       "      <td>Chr5_8901116</td>\n",
       "      <td>Col</td>\n",
       "      <td>root</td>\n",
       "      <td>Cr*Ps</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>54649</th>\n",
       "      <td>AT5G25560.4</td>\n",
       "      <td>2240</td>\n",
       "      <td>62.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>Cr*Ps r1</td>\n",
       "      <td>2240</td>\n",
       "      <td>Chr5_8901131</td>\n",
       "      <td>Col</td>\n",
       "      <td>root</td>\n",
       "      <td>Cr*Ps</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>54650</th>\n",
       "      <td>AT5G25560.4</td>\n",
       "      <td>2249</td>\n",
       "      <td>53.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>Cr*Ps r1</td>\n",
       "      <td>2249</td>\n",
       "      <td>Chr5_8901140</td>\n",
       "      <td>Col</td>\n",
       "      <td>root</td>\n",
       "      <td>Cr*Ps</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>54651</th>\n",
       "      <td>AT5G25560.4</td>\n",
       "      <td>2252</td>\n",
       "      <td>49.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>Cr*Ps r1</td>\n",
       "      <td>2252</td>\n",
       "      <td>Chr5_8901143</td>\n",
       "      <td>Col</td>\n",
       "      <td>root</td>\n",
       "      <td>Cr*Ps</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>54652</th>\n",
       "      <td>AT5G25560.4</td>\n",
       "      <td>2336</td>\n",
       "      <td>44.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>Cr*Ps r1</td>\n",
       "      <td>2336</td>\n",
       "      <td>Chr5_8901310</td>\n",
       "      <td>Col</td>\n",
       "      <td>root</td>\n",
       "      <td>Cr*Ps</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>54653 rows × 10 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "        Transcript  Informative positions      N    n       rep  snp_list  \\\n",
       "0      AT1G50920.1                    220  707.0  0.0  Cr*Ps r1       220   \n",
       "1      AT1G50920.1                   1351  597.0  0.0  Cr*Ps r1      1351   \n",
       "2      AT1G51380.1                    209   24.0  0.0  Cr*Ps r1       209   \n",
       "3      AT1G51380.1                    987    9.0  0.0  Cr*Ps r1       987   \n",
       "4      AT1G51380.1                   1376    6.0  0.0  Cr*Ps r1      1376   \n",
       "...            ...                    ...    ...  ...       ...       ...   \n",
       "54648  AT5G25560.4                   2225   61.0  0.0  Cr*Ps r1      2225   \n",
       "54649  AT5G25560.4                   2240   62.0  0.0  Cr*Ps r1      2240   \n",
       "54650  AT5G25560.4                   2249   53.0  0.0  Cr*Ps r1      2249   \n",
       "54651  AT5G25560.4                   2252   49.0  0.0  Cr*Ps r1      2252   \n",
       "54652  AT5G25560.4                   2336   44.0  0.0  Cr*Ps r1      2336   \n",
       "\n",
       "                 SNP type tissue    exp  \n",
       "0      Chr1_18870656  Col   root  Cr*Ps  \n",
       "1      Chr1_18871787  Col   root  Cr*Ps  \n",
       "2      Chr1_19048090  Col   root  Cr*Ps  \n",
       "3      Chr1_19049605  Col   root  Cr*Ps  \n",
       "4      Chr1_19050086  Col   root  Cr*Ps  \n",
       "...              ...  ...    ...    ...  \n",
       "54648   Chr5_8901116  Col   root  Cr*Ps  \n",
       "54649   Chr5_8901131  Col   root  Cr*Ps  \n",
       "54650   Chr5_8901140  Col   root  Cr*Ps  \n",
       "54651   Chr5_8901143  Col   root  Cr*Ps  \n",
       "54652   Chr5_8901310  Col   root  Cr*Ps  \n",
       "\n",
       "[54653 rows x 10 columns]"
      ]
     },
     "execution_count": 17,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df_bayes_exp[0]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Transcript</th>\n",
       "      <th>Informative positions</th>\n",
       "      <th>N</th>\n",
       "      <th>n</th>\n",
       "      <th>rep</th>\n",
       "      <th>snp_list</th>\n",
       "      <th>SNP</th>\n",
       "      <th>type</th>\n",
       "      <th>tissue</th>\n",
       "      <th>exp</th>\n",
       "      <th>condition</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>AT1G50920.1</td>\n",
       "      <td>220</td>\n",
       "      <td>707.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>Cr*Ps r1</td>\n",
       "      <td>220</td>\n",
       "      <td>Chr1_18870656</td>\n",
       "      <td>Col</td>\n",
       "      <td>root</td>\n",
       "      <td>Cr*Ps</td>\n",
       "      <td>FN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>AT1G50920.1</td>\n",
       "      <td>1351</td>\n",
       "      <td>597.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>Cr*Ps r1</td>\n",
       "      <td>1351</td>\n",
       "      <td>Chr1_18871787</td>\n",
       "      <td>Col</td>\n",
       "      <td>root</td>\n",
       "      <td>Cr*Ps</td>\n",
       "      <td>FN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>AT1G51380.1</td>\n",
       "      <td>209</td>\n",
       "      <td>24.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>Cr*Ps r1</td>\n",
       "      <td>209</td>\n",
       "      <td>Chr1_19048090</td>\n",
       "      <td>Col</td>\n",
       "      <td>root</td>\n",
       "      <td>Cr*Ps</td>\n",
       "      <td>FN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>AT1G51380.1</td>\n",
       "      <td>987</td>\n",
       "      <td>9.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>Cr*Ps r1</td>\n",
       "      <td>987</td>\n",
       "      <td>Chr1_19049605</td>\n",
       "      <td>Col</td>\n",
       "      <td>root</td>\n",
       "      <td>Cr*Ps</td>\n",
       "      <td>FN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>AT1G51380.1</td>\n",
       "      <td>1376</td>\n",
       "      <td>6.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>Cr*Ps r1</td>\n",
       "      <td>1376</td>\n",
       "      <td>Chr1_19050086</td>\n",
       "      <td>Col</td>\n",
       "      <td>root</td>\n",
       "      <td>Cr*Ps</td>\n",
       "      <td>FN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>54648</th>\n",
       "      <td>AT5G25560.4</td>\n",
       "      <td>2225</td>\n",
       "      <td>42.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>Cr*Ps -P r2</td>\n",
       "      <td>2225</td>\n",
       "      <td>Chr5_8901116</td>\n",
       "      <td>Col</td>\n",
       "      <td>root</td>\n",
       "      <td>Cr*Ps</td>\n",
       "      <td>P</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>54649</th>\n",
       "      <td>AT5G25560.4</td>\n",
       "      <td>2240</td>\n",
       "      <td>41.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>Cr*Ps -P r2</td>\n",
       "      <td>2240</td>\n",
       "      <td>Chr5_8901131</td>\n",
       "      <td>Col</td>\n",
       "      <td>root</td>\n",
       "      <td>Cr*Ps</td>\n",
       "      <td>P</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>54650</th>\n",
       "      <td>AT5G25560.4</td>\n",
       "      <td>2249</td>\n",
       "      <td>36.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>Cr*Ps -P r2</td>\n",
       "      <td>2249</td>\n",
       "      <td>Chr5_8901140</td>\n",
       "      <td>Col</td>\n",
       "      <td>root</td>\n",
       "      <td>Cr*Ps</td>\n",
       "      <td>P</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>54651</th>\n",
       "      <td>AT5G25560.4</td>\n",
       "      <td>2252</td>\n",
       "      <td>38.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>Cr*Ps -P r2</td>\n",
       "      <td>2252</td>\n",
       "      <td>Chr5_8901143</td>\n",
       "      <td>Col</td>\n",
       "      <td>root</td>\n",
       "      <td>Cr*Ps</td>\n",
       "      <td>P</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>54652</th>\n",
       "      <td>AT5G25560.4</td>\n",
       "      <td>2336</td>\n",
       "      <td>47.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>Cr*Ps -P r2</td>\n",
       "      <td>2336</td>\n",
       "      <td>Chr5_8901310</td>\n",
       "      <td>Col</td>\n",
       "      <td>root</td>\n",
       "      <td>Cr*Ps</td>\n",
       "      <td>P</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>1584937 rows × 11 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "        Transcript  Informative positions      N    n          rep  snp_list  \\\n",
       "0      AT1G50920.1                    220  707.0  0.0     Cr*Ps r1       220   \n",
       "1      AT1G50920.1                   1351  597.0  0.0     Cr*Ps r1      1351   \n",
       "2      AT1G51380.1                    209   24.0  0.0     Cr*Ps r1       209   \n",
       "3      AT1G51380.1                    987    9.0  0.0     Cr*Ps r1       987   \n",
       "4      AT1G51380.1                   1376    6.0  0.0     Cr*Ps r1      1376   \n",
       "...            ...                    ...    ...  ...          ...       ...   \n",
       "54648  AT5G25560.4                   2225   42.0  0.0  Cr*Ps -P r2      2225   \n",
       "54649  AT5G25560.4                   2240   41.0  0.0  Cr*Ps -P r2      2240   \n",
       "54650  AT5G25560.4                   2249   36.0  0.0  Cr*Ps -P r2      2249   \n",
       "54651  AT5G25560.4                   2252   38.0  0.0  Cr*Ps -P r2      2252   \n",
       "54652  AT5G25560.4                   2336   47.0  0.0  Cr*Ps -P r2      2336   \n",
       "\n",
       "                 SNP type tissue    exp condition  \n",
       "0      Chr1_18870656  Col   root  Cr*Ps        FN  \n",
       "1      Chr1_18871787  Col   root  Cr*Ps        FN  \n",
       "2      Chr1_19048090  Col   root  Cr*Ps        FN  \n",
       "3      Chr1_19049605  Col   root  Cr*Ps        FN  \n",
       "4      Chr1_19050086  Col   root  Cr*Ps        FN  \n",
       "...              ...  ...    ...    ...       ...  \n",
       "54648   Chr5_8901116  Col   root  Cr*Ps         P  \n",
       "54649   Chr5_8901131  Col   root  Cr*Ps         P  \n",
       "54650   Chr5_8901140  Col   root  Cr*Ps         P  \n",
       "54651   Chr5_8901143  Col   root  Cr*Ps         P  \n",
       "54652   Chr5_8901310  Col   root  Cr*Ps         P  \n",
       "\n",
       "[1584937 rows x 11 columns]"
      ]
     },
     "execution_count": 18,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df_bayes = pd.concat(df_bayes_exp)\n",
    "## Need to separate by experiment type\n",
    "df_bayes[\"condition\"] = \"FN\"\n",
    "df_bayes.loc[df_bayes[\"rep\"].str.contains(\"-N\"),\"condition\"] = \"N\"\n",
    "df_bayes.loc[df_bayes[\"rep\"].str.contains(\"-P\"),\"condition\"] = \"P\"\n",
    "df_bayes\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>SNP</th>\n",
       "      <th>condition</th>\n",
       "      <th>exp</th>\n",
       "      <th>Transcript</th>\n",
       "      <th>tissue</th>\n",
       "      <th>type</th>\n",
       "      <th>N</th>\n",
       "      <th>n</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>Chr1_1000114</td>\n",
       "      <td>FN</td>\n",
       "      <td>Cr*Ps</td>\n",
       "      <td>AT1G03910.1</td>\n",
       "      <td>root</td>\n",
       "      <td>Col</td>\n",
       "      <td>178.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>Chr1_1000114</td>\n",
       "      <td>FN</td>\n",
       "      <td>Pr*Cs</td>\n",
       "      <td>AT1G03910.1</td>\n",
       "      <td>root</td>\n",
       "      <td>Ped</td>\n",
       "      <td>187.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>Chr1_1000114</td>\n",
       "      <td>FN</td>\n",
       "      <td>Pr*Cs FN_root</td>\n",
       "      <td>AT1G03910.1</td>\n",
       "      <td>root</td>\n",
       "      <td>Ped</td>\n",
       "      <td>235.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>Chr1_1000114</td>\n",
       "      <td>FN</td>\n",
       "      <td>PrCs*</td>\n",
       "      <td>AT1G03910.1</td>\n",
       "      <td>shoot</td>\n",
       "      <td>Col</td>\n",
       "      <td>133.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>Chr1_1000114</td>\n",
       "      <td>FN</td>\n",
       "      <td>PrCs* FN_flower</td>\n",
       "      <td>AT1G03910.1</td>\n",
       "      <td>shoot</td>\n",
       "      <td>Col</td>\n",
       "      <td>223.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>584397</th>\n",
       "      <td>Chr5_9987837</td>\n",
       "      <td>FN</td>\n",
       "      <td>CrPs*</td>\n",
       "      <td>AT5G27950.1</td>\n",
       "      <td>shoot</td>\n",
       "      <td>Ped</td>\n",
       "      <td>25.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>584398</th>\n",
       "      <td>Chr5_9987837</td>\n",
       "      <td>N</td>\n",
       "      <td>Cr*Ps</td>\n",
       "      <td>AT5G27950.1</td>\n",
       "      <td>root</td>\n",
       "      <td>Col</td>\n",
       "      <td>43.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>584399</th>\n",
       "      <td>Chr5_9987837</td>\n",
       "      <td>N</td>\n",
       "      <td>CrPs*</td>\n",
       "      <td>AT5G27950.1</td>\n",
       "      <td>shoot</td>\n",
       "      <td>Ped</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>584400</th>\n",
       "      <td>Chr5_9987837</td>\n",
       "      <td>P</td>\n",
       "      <td>Cr*Ps</td>\n",
       "      <td>AT5G27950.1</td>\n",
       "      <td>root</td>\n",
       "      <td>Col</td>\n",
       "      <td>32.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>584401</th>\n",
       "      <td>Chr5_9987837</td>\n",
       "      <td>P</td>\n",
       "      <td>CrPs*</td>\n",
       "      <td>AT5G27950.1</td>\n",
       "      <td>shoot</td>\n",
       "      <td>Ped</td>\n",
       "      <td>17.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>584402 rows × 8 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                 SNP condition              exp   Transcript tissue type  \\\n",
       "0       Chr1_1000114        FN            Cr*Ps  AT1G03910.1   root  Col   \n",
       "1       Chr1_1000114        FN            Pr*Cs  AT1G03910.1   root  Ped   \n",
       "2       Chr1_1000114        FN    Pr*Cs FN_root  AT1G03910.1   root  Ped   \n",
       "3       Chr1_1000114        FN            PrCs*  AT1G03910.1  shoot  Col   \n",
       "4       Chr1_1000114        FN  PrCs* FN_flower  AT1G03910.1  shoot  Col   \n",
       "...              ...       ...              ...          ...    ...  ...   \n",
       "584397  Chr5_9987837        FN            CrPs*  AT5G27950.1  shoot  Ped   \n",
       "584398  Chr5_9987837         N            Cr*Ps  AT5G27950.1   root  Col   \n",
       "584399  Chr5_9987837         N            CrPs*  AT5G27950.1  shoot  Ped   \n",
       "584400  Chr5_9987837         P            Cr*Ps  AT5G27950.1   root  Col   \n",
       "584401  Chr5_9987837         P            CrPs*  AT5G27950.1  shoot  Ped   \n",
       "\n",
       "            N    n  \n",
       "0       178.0  0.0  \n",
       "1       187.0  0.0  \n",
       "2       235.0  0.0  \n",
       "3       133.0  0.0  \n",
       "4       223.0  0.0  \n",
       "...       ...  ...  \n",
       "584397   25.0  0.0  \n",
       "584398   43.0  0.0  \n",
       "584399    0.0  0.0  \n",
       "584400   32.0  0.0  \n",
       "584401   17.0  0.0  \n",
       "\n",
       "[584402 rows x 8 columns]"
      ]
     },
     "execution_count": 19,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "\n",
    "df_bayes_grouped = df_bayes.groupby([\"SNP\",\"condition\",\"tissue\",\"type\",\"exp\"], as_index=False).agg({\"Transcript\": \"first\", \"tissue\":\"first\",\"type\":\"first\",\"N\": \"sum\", \"n\": \"sum\"})\n",
    "df_bayes_grouped"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "{'Cr*Ps',\n",
       " 'CrPs*',\n",
       " 'Pr*Cs',\n",
       " 'Pr*Cs FN_root',\n",
       " 'PrCs*',\n",
       " 'PrCs* FN_flower',\n",
       " 'PrCs* FN_rosette',\n",
       " 'PrCs* FN_stemLower',\n",
       " 'PrCs* FN_stemUpper'}"
      ]
     },
     "execution_count": 20,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "set(df_bayes_grouped[\"exp\"].to_list())"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>SNP</th>\n",
       "      <th>condition</th>\n",
       "      <th>exp</th>\n",
       "      <th>Transcript</th>\n",
       "      <th>tissue</th>\n",
       "      <th>type</th>\n",
       "      <th>N</th>\n",
       "      <th>n</th>\n",
       "      <th>Nh1</th>\n",
       "      <th>nh1</th>\n",
       "      <th>Nh2</th>\n",
       "      <th>nh2</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>Chr1_1000114</td>\n",
       "      <td>FN</td>\n",
       "      <td>Cr*Ps</td>\n",
       "      <td>AT1G03910.1</td>\n",
       "      <td>root</td>\n",
       "      <td>Col</td>\n",
       "      <td>178.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>201.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>125.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>Chr1_1000114</td>\n",
       "      <td>N</td>\n",
       "      <td>Cr*Ps</td>\n",
       "      <td>AT1G03910.1</td>\n",
       "      <td>root</td>\n",
       "      <td>Col</td>\n",
       "      <td>117.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>201.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>125.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>Chr1_1000114</td>\n",
       "      <td>P</td>\n",
       "      <td>Cr*Ps</td>\n",
       "      <td>AT1G03910.1</td>\n",
       "      <td>root</td>\n",
       "      <td>Col</td>\n",
       "      <td>84.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>201.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>125.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>Chr1_10010338</td>\n",
       "      <td>FN</td>\n",
       "      <td>Cr*Ps</td>\n",
       "      <td>AT1G28470.1</td>\n",
       "      <td>root</td>\n",
       "      <td>Col</td>\n",
       "      <td>153.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>75.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>43.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>Chr1_10010338</td>\n",
       "      <td>N</td>\n",
       "      <td>Cr*Ps</td>\n",
       "      <td>AT1G28470.1</td>\n",
       "      <td>root</td>\n",
       "      <td>Col</td>\n",
       "      <td>402.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>75.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>43.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>125974</th>\n",
       "      <td>Chr5_9987395</td>\n",
       "      <td>N</td>\n",
       "      <td>CrPs*</td>\n",
       "      <td>AT5G27950.1</td>\n",
       "      <td>shoot</td>\n",
       "      <td>Ped</td>\n",
       "      <td>286.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>76.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>316.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>125975</th>\n",
       "      <td>Chr5_9987395</td>\n",
       "      <td>P</td>\n",
       "      <td>CrPs*</td>\n",
       "      <td>AT5G27950.1</td>\n",
       "      <td>shoot</td>\n",
       "      <td>Ped</td>\n",
       "      <td>205.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>76.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>316.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>125976</th>\n",
       "      <td>Chr5_9987837</td>\n",
       "      <td>FN</td>\n",
       "      <td>CrPs*</td>\n",
       "      <td>AT5G27950.1</td>\n",
       "      <td>shoot</td>\n",
       "      <td>Ped</td>\n",
       "      <td>25.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>15.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>52.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>125977</th>\n",
       "      <td>Chr5_9987837</td>\n",
       "      <td>N</td>\n",
       "      <td>CrPs*</td>\n",
       "      <td>AT5G27950.1</td>\n",
       "      <td>shoot</td>\n",
       "      <td>Ped</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>15.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>52.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>125978</th>\n",
       "      <td>Chr5_9987837</td>\n",
       "      <td>P</td>\n",
       "      <td>CrPs*</td>\n",
       "      <td>AT5G27950.1</td>\n",
       "      <td>shoot</td>\n",
       "      <td>Ped</td>\n",
       "      <td>17.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>15.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>52.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>552469 rows × 12 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                  SNP condition    exp   Transcript tissue type      N    n  \\\n",
       "0        Chr1_1000114        FN  Cr*Ps  AT1G03910.1   root  Col  178.0  0.0   \n",
       "1        Chr1_1000114         N  Cr*Ps  AT1G03910.1   root  Col  117.0  0.0   \n",
       "2        Chr1_1000114         P  Cr*Ps  AT1G03910.1   root  Col   84.0  0.0   \n",
       "3       Chr1_10010338        FN  Cr*Ps  AT1G28470.1   root  Col  153.0  0.0   \n",
       "4       Chr1_10010338         N  Cr*Ps  AT1G28470.1   root  Col  402.0  0.0   \n",
       "...               ...       ...    ...          ...    ...  ...    ...  ...   \n",
       "125974   Chr5_9987395         N  CrPs*  AT5G27950.1  shoot  Ped  286.0  0.0   \n",
       "125975   Chr5_9987395         P  CrPs*  AT5G27950.1  shoot  Ped  205.0  0.0   \n",
       "125976   Chr5_9987837        FN  CrPs*  AT5G27950.1  shoot  Ped   25.0  0.0   \n",
       "125977   Chr5_9987837         N  CrPs*  AT5G27950.1  shoot  Ped    0.0  0.0   \n",
       "125978   Chr5_9987837         P  CrPs*  AT5G27950.1  shoot  Ped   17.0  0.0   \n",
       "\n",
       "          Nh1  nh1    Nh2  nh2  \n",
       "0       201.0  0.0  125.0  0.0  \n",
       "1       201.0  0.0  125.0  0.0  \n",
       "2       201.0  0.0  125.0  0.0  \n",
       "3        75.0  0.0   43.0  0.0  \n",
       "4        75.0  0.0   43.0  0.0  \n",
       "...       ...  ...    ...  ...  \n",
       "125974   76.0  0.0  316.0  0.0  \n",
       "125975   76.0  0.0  316.0  0.0  \n",
       "125976   15.0  0.0   52.0  0.0  \n",
       "125977   15.0  0.0   52.0  0.0  \n",
       "125978   15.0  0.0   52.0  0.0  \n",
       "\n",
       "[552469 rows x 12 columns]"
      ]
     },
     "execution_count": 21,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "## Time to add it all together!\n",
    "\n",
    "df_bayes_croot = df_bayes_grouped[(df_bayes_grouped[\"tissue\"]==\"root\") & (df_bayes_grouped[\"type\"]==\"Col\")]\n",
    "df_bayes_proot = df_bayes_grouped[(df_bayes_grouped[\"tissue\"]==\"root\") & (df_bayes_grouped[\"type\"]==\"Ped\")]\n",
    "df_bayes_cshoot = df_bayes_grouped[(df_bayes_grouped[\"tissue\"]==\"shoot\") & (df_bayes_grouped[\"type\"]==\"Col\")]\n",
    "df_bayes_pshoot = df_bayes_grouped[(df_bayes_grouped[\"tissue\"]==\"shoot\") & (df_bayes_grouped[\"type\"]==\"Ped\")]\n",
    "\n",
    "hom1 = ccroot.copy()\n",
    "hom2 = pproot.copy()\n",
    "hom1.rename(columns = {\"depth\":\"Nh1\",\"lerDepth\":\"nh1\"}, inplace=True)\n",
    "hom2.rename(columns = {\"depth\":\"Nh2\",\"colDepth\":\"nh2\"}, inplace=True)\n",
    "df_bayes_croot = df_bayes_croot.merge(hom1[[\"SNP\",\"Nh1\",\"nh1\"]], on = \"SNP\")\n",
    "df_bayes_croot = df_bayes_croot.merge(hom2[[\"SNP\",\"Nh2\",\"nh2\"]], on = \"SNP\")\n",
    "\n",
    "hom1 = pproot.copy()\n",
    "hom2 = ccroot.copy()\n",
    "hom1.rename(columns = {\"depth\":\"Nh1\",\"colDepth\":\"nh1\"}, inplace=True)\n",
    "hom2.rename(columns = {\"depth\":\"Nh2\",\"lerDepth\":\"nh2\"}, inplace=True)\n",
    "df_bayes_proot = df_bayes_proot.merge(hom1[[\"SNP\",\"Nh1\",\"nh1\"]], on = \"SNP\")\n",
    "df_bayes_proot = df_bayes_proot.merge(hom2[[\"SNP\",\"Nh2\",\"nh2\"]], on = \"SNP\")\n",
    "\n",
    "hom1 = ccshoot.copy()\n",
    "hom2 = ppshoot.copy()\n",
    "hom1.rename(columns = {\"depth\":\"Nh1\",\"lerDepth\":\"nh1\"}, inplace=True)\n",
    "hom2.rename(columns = {\"depth\":\"Nh2\",\"colDepth\":\"nh2\"}, inplace=True)\n",
    "df_bayes_cshoot = df_bayes_cshoot.merge(hom1[[\"SNP\",\"Nh1\",\"nh1\"]], on = \"SNP\")\n",
    "df_bayes_cshoot = df_bayes_cshoot.merge(hom2[[\"SNP\",\"Nh2\",\"nh2\"]], on = \"SNP\")\n",
    "\n",
    "hom1 = ppshoot.copy()\n",
    "hom2 = ccshoot.copy()\n",
    "hom1.rename(columns = {\"depth\":\"Nh1\",\"colDepth\":\"nh1\"}, inplace=True)\n",
    "hom2.rename(columns = {\"depth\":\"Nh2\",\"lerDepth\":\"nh2\"}, inplace=True)\n",
    "df_bayes_pshoot = df_bayes_pshoot.merge(hom1[[\"SNP\",\"Nh1\",\"nh1\"]], on = \"SNP\")\n",
    "df_bayes_pshoot = df_bayes_pshoot.merge(hom2[[\"SNP\",\"Nh2\",\"nh2\"]], on = \"SNP\")\n",
    "\n",
    "df_bayes_final = pd.concat([df_bayes_croot, df_bayes_proot, df_bayes_cshoot, df_bayes_pshoot])\n",
    "df_bayes_final"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 25,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/Users/tomkinsm/miniconda3/lib/python3.10/site-packages/baymobil/baymobil.py:94: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame.\n",
      "Try using .loc[row_indexer,col_indexer] = value instead\n",
      "\n",
      "See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n",
      "  df[cols] = df[cols].apply(pd.to_numeric, errors='coerce')\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Warning: 17333 Zero values were removed.\n",
      "INFO: Pandarallel will run on 10 workers.\n",
      "INFO: Pandarallel will use standard multiprocessing data transfer (pipe) to transfer data between the main process and workers.\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/Users/tomkinsm/miniconda3/lib/python3.10/site-packages/baymobil/baymobil.py:96: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame\n",
      "\n",
      "See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n",
      "  df.dropna(inplace=True)\n"
     ]
    },
    {
     "data": {
      "application/vnd.jupyter.widget-view+json": {
       "model_id": "f03950e0af9a47e9a46c2e60bf08dcc6",
       "version_major": 2,
       "version_minor": 0
      },
      "text/plain": [
       "VBox(children=(HBox(children=(IntProgress(value=0, description='0.00%', max=53482), Label(value='0 / 53482')))…"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "import baymobil as baymob\n",
    "\n",
    "## WE seem to have some dodgy values where n is greater than N - maybe check those out\n",
    "df_bayes_final = df_bayes_final[df_bayes_final[\"n\"] <= df_bayes_final[\"N\"]]\n",
    "df_results = baymob.run_bayes(df_bayes_final, 10)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 26,
   "metadata": {},
   "outputs": [],
   "source": [
    "df_results.to_csv(\"thieme_new_analysis.csv\", index=None)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 31,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/var/folders/t5/q1tt12hd2sg29kw894zcfbtm000b_5/T/ipykernel_91002/2334272700.py:5: FutureWarning: The default value of numeric_only in DataFrameGroupBy.sum is deprecated. In a future version, numeric_only will default to False. Either specify numeric_only or select only columns which should be valid for the function.\n",
      "  df_results_grouped = df_results.groupby([\"Transcript\",\"condition\",\"exp\",\"type\",\"tissue\"]).sum().reset_index()\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Transcript</th>\n",
       "      <th>condition</th>\n",
       "      <th>exp</th>\n",
       "      <th>type</th>\n",
       "      <th>tissue</th>\n",
       "      <th>N</th>\n",
       "      <th>n</th>\n",
       "      <th>Nh1</th>\n",
       "      <th>nh1</th>\n",
       "      <th>Nh2</th>\n",
       "      <th>nh2</th>\n",
       "      <th>nmax</th>\n",
       "      <th>meanN2</th>\n",
       "      <th>N2max</th>\n",
       "      <th>log10BF</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>2067</th>\n",
       "      <td>AT1G06680.1</td>\n",
       "      <td>FN</td>\n",
       "      <td>Pr*Cs FN_root</td>\n",
       "      <td>Ped</td>\n",
       "      <td>root</td>\n",
       "      <td>19.0</td>\n",
       "      <td>10.0</td>\n",
       "      <td>80.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>93.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>10</td>\n",
       "      <td>9.966324</td>\n",
       "      <td>10</td>\n",
       "      <td>2.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3258</th>\n",
       "      <td>AT1G09560.1</td>\n",
       "      <td>FN</td>\n",
       "      <td>Pr*Cs FN_root</td>\n",
       "      <td>Ped</td>\n",
       "      <td>root</td>\n",
       "      <td>29922.0</td>\n",
       "      <td>83.0</td>\n",
       "      <td>22715.0</td>\n",
       "      <td>11.0</td>\n",
       "      <td>13272.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>90</td>\n",
       "      <td>57.131826</td>\n",
       "      <td>70</td>\n",
       "      <td>4.667092</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3303</th>\n",
       "      <td>AT1G09640.1</td>\n",
       "      <td>FN</td>\n",
       "      <td>Pr*Cs</td>\n",
       "      <td>Ped</td>\n",
       "      <td>root</td>\n",
       "      <td>2123.0</td>\n",
       "      <td>16.0</td>\n",
       "      <td>1569.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>2389.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>10</td>\n",
       "      <td>12.588884</td>\n",
       "      <td>15</td>\n",
       "      <td>1.481299</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4080</th>\n",
       "      <td>AT1G11580.1</td>\n",
       "      <td>FN</td>\n",
       "      <td>PrCs* FN_rosette</td>\n",
       "      <td>Col</td>\n",
       "      <td>shoot</td>\n",
       "      <td>2289.0</td>\n",
       "      <td>43.0</td>\n",
       "      <td>3515.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1785.0</td>\n",
       "      <td>2.0</td>\n",
       "      <td>60</td>\n",
       "      <td>36.237985</td>\n",
       "      <td>43</td>\n",
       "      <td>7.475608</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4299</th>\n",
       "      <td>AT1G12010.1</td>\n",
       "      <td>FN</td>\n",
       "      <td>PrCs* FN_rosette</td>\n",
       "      <td>Col</td>\n",
       "      <td>shoot</td>\n",
       "      <td>53.0</td>\n",
       "      <td>8.0</td>\n",
       "      <td>405.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>3.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>30</td>\n",
       "      <td>12.093939</td>\n",
       "      <td>9</td>\n",
       "      <td>3.114719</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>88666</th>\n",
       "      <td>AT5G65220.1</td>\n",
       "      <td>FN</td>\n",
       "      <td>Pr*Cs FN_root</td>\n",
       "      <td>Ped</td>\n",
       "      <td>root</td>\n",
       "      <td>153.0</td>\n",
       "      <td>6.0</td>\n",
       "      <td>179.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>209.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>10</td>\n",
       "      <td>4.800331</td>\n",
       "      <td>6</td>\n",
       "      <td>1.198098</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>89217</th>\n",
       "      <td>AT5G66570.1</td>\n",
       "      <td>FN</td>\n",
       "      <td>Pr*Cs FN_root</td>\n",
       "      <td>Ped</td>\n",
       "      <td>root</td>\n",
       "      <td>354.0</td>\n",
       "      <td>140.0</td>\n",
       "      <td>204.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>603.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>20</td>\n",
       "      <td>137.047010</td>\n",
       "      <td>139</td>\n",
       "      <td>4.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>89223</th>\n",
       "      <td>AT5G66570.1</td>\n",
       "      <td>N</td>\n",
       "      <td>Cr*Ps</td>\n",
       "      <td>Col</td>\n",
       "      <td>root</td>\n",
       "      <td>349.0</td>\n",
       "      <td>12.0</td>\n",
       "      <td>603.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>204.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>20</td>\n",
       "      <td>10.610598</td>\n",
       "      <td>12</td>\n",
       "      <td>3.076923</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>89225</th>\n",
       "      <td>AT5G66570.1</td>\n",
       "      <td>P</td>\n",
       "      <td>Cr*Ps</td>\n",
       "      <td>Col</td>\n",
       "      <td>root</td>\n",
       "      <td>171.0</td>\n",
       "      <td>15.0</td>\n",
       "      <td>603.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>204.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>20</td>\n",
       "      <td>14.589373</td>\n",
       "      <td>15</td>\n",
       "      <td>4.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>89616</th>\n",
       "      <td>AT5G67590.1</td>\n",
       "      <td>FN</td>\n",
       "      <td>Pr*Cs FN_root</td>\n",
       "      <td>Ped</td>\n",
       "      <td>root</td>\n",
       "      <td>614.0</td>\n",
       "      <td>7.0</td>\n",
       "      <td>809.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>54.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>10</td>\n",
       "      <td>6.054679</td>\n",
       "      <td>7</td>\n",
       "      <td>1.632523</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>164 rows × 15 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "        Transcript condition               exp type tissue        N      n  \\\n",
       "2067   AT1G06680.1        FN     Pr*Cs FN_root  Ped   root     19.0   10.0   \n",
       "3258   AT1G09560.1        FN     Pr*Cs FN_root  Ped   root  29922.0   83.0   \n",
       "3303   AT1G09640.1        FN             Pr*Cs  Ped   root   2123.0   16.0   \n",
       "4080   AT1G11580.1        FN  PrCs* FN_rosette  Col  shoot   2289.0   43.0   \n",
       "4299   AT1G12010.1        FN  PrCs* FN_rosette  Col  shoot     53.0    8.0   \n",
       "...            ...       ...               ...  ...    ...      ...    ...   \n",
       "88666  AT5G65220.1        FN     Pr*Cs FN_root  Ped   root    153.0    6.0   \n",
       "89217  AT5G66570.1        FN     Pr*Cs FN_root  Ped   root    354.0  140.0   \n",
       "89223  AT5G66570.1         N             Cr*Ps  Col   root    349.0   12.0   \n",
       "89225  AT5G66570.1         P             Cr*Ps  Col   root    171.0   15.0   \n",
       "89616  AT5G67590.1        FN     Pr*Cs FN_root  Ped   root    614.0    7.0   \n",
       "\n",
       "           Nh1   nh1      Nh2  nh2  nmax      meanN2  N2max   log10BF  \n",
       "2067      80.0   0.0     93.0  0.0    10    9.966324     10  2.000000  \n",
       "3258   22715.0  11.0  13272.0  1.0    90   57.131826     70  4.667092  \n",
       "3303    1569.0   1.0   2389.0  1.0    10   12.588884     15  1.481299  \n",
       "4080    3515.0   1.0   1785.0  2.0    60   36.237985     43  7.475608  \n",
       "4299     405.0   0.0      3.0  0.0    30   12.093939      9  3.114719  \n",
       "...        ...   ...      ...  ...   ...         ...    ...       ...  \n",
       "88666    179.0   0.0    209.0  0.0    10    4.800331      6  1.198098  \n",
       "89217    204.0   1.0    603.0  0.0    20  137.047010    139  4.000000  \n",
       "89223    603.0   0.0    204.0  1.0    20   10.610598     12  3.076923  \n",
       "89225    603.0   0.0    204.0  1.0    20   14.589373     15  4.000000  \n",
       "89616    809.0   0.0     54.0  0.0    10    6.054679      7  1.632523  \n",
       "\n",
       "[164 rows x 15 columns]"
      ]
     },
     "execution_count": 31,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "## Cap the log10BF in df_results between -2 and 2\n",
    "df_results[\"log10BF\"] = df_results[\"log10BF\"].apply(lambda x: 2 if x > 2 else x)\n",
    "df_results[\"log10BF\"] = df_results[\"log10BF\"].apply(lambda x: -2 if x < -2 else x)\n",
    "\n",
    "df_results_grouped = df_results.groupby([\"Transcript\",\"condition\",\"exp\",\"type\",\"tissue\"]).sum().reset_index()\n",
    "df_results_grouped[df_results_grouped[\"log10BF\"]>=1]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.10.8"
  },
  "orig_nbformat": 4
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
